\documentclass{article}
\input{../config/commontex}

\title{The query language}

\author{Lobry, J.R.}

\usepackage{/Library/Frameworks/R.framework/Resources/share/texmf/Sweave}
\begin{document}
%
% To change the R input/output style:
%
\definecolor{Soutput}{rgb}{0,0,0.56}
\definecolor{Sinput}{rgb}{0.56,0,0}
\DefineVerbatimEnvironment{Sinput}{Verbatim}
{formatcom={\color{Sinput}},fontsize=\footnotesize, baselinestretch=0.75}
\DefineVerbatimEnvironment{Soutput}{Verbatim}
{formatcom={\color{Soutput}},fontsize=\footnotesize, baselinestretch=0.75}
%
% This removes the extra spacing after code and output chunks in Sweave,
% but keeps the spacing around the whole block.
%
\fvset{listparameters={\setlength{\topsep}{0pt}}}
\renewenvironment{Schunk}{\vspace{\topsep}}{\vspace{\topsep}}
%
% Rlogo
%
\newcommand{\Rlogo}{\protect\includegraphics[height=1.8ex,keepaspectratio]{../figs/Rlogo.pdf}}
%
% Shortcut for seqinR:
%
\newcommand{\seqinr}{\texttt{seqin\bf{R}}}
\newcommand{\Seqinr}{\texttt{Seqin\bf{R}}}
\fvset{fontsize= \scriptsize}
%
% R output options and libraries to be loaded.
%
%
%  Sweave Options
%
% Put all figures in the fig folder and start the name with current file name.
% Do not produce EPS files
%


\maketitle
\tableofcontents
% BEGIN - DO NOT REMOVE THIS LINE
\label{querylang}

\section{Where to find information}

The last version of the documentation for the query language is available online
at \url{http://pbil.univ-lyon1.fr/databases/acnuc/cfonctions.html#QUERYLANGUAGE}.
This documentation has been imported within the documentation of the
\texttt{query()} function, but the last available update is the online version.
The query language is a specificity of the ACNUC system \cite{acnuc1984, acnuc1985,
acnuc1985b, acnuc2007}.

\section{Case sensitivity and ambiguities resolution}

The query language is case insensitive, for instance:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("lowercase", "sp=escherichia coli", virtual = TRUE)
 query("uppercase", "SP=Escherichia coli", virtual = TRUE)
 lowercase$nelem == uppercase$nelem
\end{Sinput}
\begin{Soutput}
[1] TRUE
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

Three operators (AND, OR, NOT) 
can be ambiguous because they can also occur within valid criterion values. 
Such ambiguities can be solved by encapsulating elementary selection 
criteria between escaped double quotes. For example:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("ambig", "\"sp=Beak and feather disease virus\" AND \"au=ritchie\"", 
     virtual = T)
 ambig$nelem
\end{Sinput}
\begin{Soutput}
[1] 18
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\section{Selection criteria}

\subsection{Introduction}

Selection criteria are in the form \texttt{c=something} (without space
before the = sign) or \texttt{list\_name} where \texttt{list\_name} is a
previously constructed list.

\subsection{\texttt{SP=taxon}}

This is used to select sequences attached to a given taxon or any other below in
the tree. The at sign @ substitutes as a wildcard character for any zero or more 
characters. Here are some examples:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("bb", "sp=Borrelia burgdorferi", virtual = T)
 bb$nelem
\end{Sinput}
\begin{Soutput}
[1] 1682
\end{Soutput}
\begin{Sinput}
 query("borrelia", "sp=Borrelia", virtual = T)
 borrelia$nelem
\end{Sinput}
\begin{Soutput}
[1] 3173
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

Here is an example of use of the wildcard @ to look for sapiens species:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("sapiens", "sp=@sapiens@", virtual = T)
 sapiens$nelem
\end{Sinput}
\begin{Soutput}
[1] 2216556
\end{Soutput}
\begin{Sinput}
 query("sapienspecies", "PS sapiens")
 getName(sapienspecies)
\end{Sinput}
\begin{Soutput}
 [1] "HOMO SAPIENS"                            
 [2] "HOMO SAPIENS NEANDERTHALENSIS"           
 [3] "HOMO SAPIENS X HUMAN PAPILLOMAVIRUS TYPE"
 [4] "HOMO SAPIENS X SIMIAN VIRUS 40"          
 [5] "HOMO SAPIENS X HUMAN ENDOGENOUS RETROVIR"
 [6] "HOMO SAPIENS X HUMAN T-CELL LYMPHOTROPIC"
 [7] "HEPATITIS B VIRUS X HOMO SAPIENS"        
 [8] "HOMO SAPIENS X HEPATITIS B VIRUS"        
 [9] "HOMO SAPIENS X HUMAN IMMUNODEFICIENCY VI"
[10] "SYNTHETIC CONSTRUCT X HOMO SAPIENS"      
[11] "HUMAN PAPILLOMAVIRUS X HOMO SAPIENS"     
[12] "MUS SP. X HOMO SAPIENS"                  
[13] "HOMO SAPIENS X HUMAN PAPILLOMAVIRUS"     
[14] "HOMO SAPIENS X HUMAN ADENOVIRUS TYPE 5"  
[15] "HOMO SAPIENS X HERV-H/ENV62"             
[16] "HOMO SAPIENS X HERV-H/ENV60"             
[17] "HOMO SAPIENS X HERV-H/ENV59"             
[18] "EXPRESSION VECTOR PTH-HIN X HOMO SAPIENS"
[19] "ADENO-ASSOCIATED VIRUS 2 X HOMO SAPIENS" 
[20] "SIMIAN VIRUS 40 X HOMO SAPIENS"          
[21] "HOMO SAPIENS X MUS MUSCULUS"             
[22] "HOMO SAPIENS X INFLUENZA B VIRUS (B/LEE/"
[23] "MUS MUSCULUS X HOMO SAPIENS"             
[24] "CRICETULUS GRISEUS X HOMO SAPIENS"       
[25] "TRYPANOSOMA CRUZI X HOMO SAPIENS"        
[26] "HOMO SAPIENS X TRYPANOSOMA CRUZI"        
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{TID=id}}

This is used to select sequences attached attached to a given numerical 
NCBI's taxonomy ID. For instance, the taxonomy ID for 
\textit{Homo sapiens neanderthalensis} is 63221:

\marginpar{\includegraphics[width=\marginparwidth]{../figs/hsn}\\
\tiny{\textit{Homo neanderthalensis}. Source: wikipedia}}

\begin{Schunk}
\begin{Sinput}
 choosebank("genbank")
 query("hsn", "TID=63221", virtual = T)
 hsn$nelem
\end{Sinput}
\begin{Soutput}
[1] 1339
\end{Soutput}
\begin{Sinput}
 query("hsnsp", "PS hsn")
 getName(hsnsp)
\end{Sinput}
\begin{Soutput}
[1] "HOMO SAPIENS NEANDERTHALENSIS"
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{K=keyword}}

This is used to select sequences attached to a given keyword or any other below in
the tree. The at sign @ substitutes as a wildcard character for any zero or more 
characters. Example:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("ecoliribprot", "sp=escherichia coli AND k=rib@ prot@", 
     virtual = T)
 ecoliribprot$nelem
\end{Sinput}
\begin{Soutput}
[1] 105
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{T=type}}

This is used to select sequences of specified type. The list of available
type for the currently opened database is given by function \texttt{getType()}:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 getType()
\end{Sinput}
\begin{Soutput}
        sname                                  libel
2661      CDS              .PE protein coding region
2662       ID                            Locus entry
2663 MISC_RNA .RN other structural RNA coding region
2664     RRNA          .RR Ribosomal RNA coding gene
2665    SCRNA              .SC small cytoplasmic RNA
2666    SNRNA                  .SN small nuclear RNA
2667     TRNA           .TR Transfer RNA coding gene
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

For instance, to select all coding sequences from \textit{Homo sapiens} we can use:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("hscds", "sp=Homo sapiens AND t=cds", virtual = T)
 hscds$nelem
\end{Sinput}
\begin{Soutput}
[1] 150513
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{J=journal\_name}}

This is used to select sequences published in journal specified using defined journal code.
For instance to select all sequences published in \textit{Science}:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("allseqsfromscience", "J=Science", virtual = TRUE)
 allseqsfromscience$nelem
\end{Sinput}
\begin{Soutput}
[1] 930397
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

The list of available journal code can be obtained from the \texttt{readsmj()} function
this way:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 nl <- readfirstrec(type = "SMJ")
 smj <- readsmj(nl = nl, all.add = TRUE)
 head(smj[!is.na(smj$nature) & smj$nature == "journal", c("sname", 
     "libel")])
\end{Sinput}
\begin{Soutput}
                sname                                                libel
21                ABP                                   Acta Biochim. Pol.
22  ABSTR-SOCNEUROSCI                              Abstr. - Soc. Neurosci.
23 ABSTRGENMEETAMSOCM                Abstr. Gen. Meet. Am. Soc. Microbiol.
24 ABSTRMIDWINTERRESM Abstr. Midwinter Res. Meet. Assoc. Res. Otolaryngol.
25 ACTAAGRICSCANDAANI                      Acta Agric. Scand. A Anim. Sci.
26 ACTABIOCHIMBIOPHYS                          Acta Biochim. Biophys. Sin.
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{R=refcode}}

This is used to select sequences from a given bibliographical reference specified
as \texttt{jcode/volume/page}. For instance, to select sequences associated
with the first publication \cite{jrl0009} of the complete genome of \textit{Rickettsia prowazekii},
we can use:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("rpro", "R=Nature/396/133")
 getName(rpro)
\end{Sinput}
\begin{Soutput}
[1] "RPDNAOMPB" "RPXX01"    "RPXX02"    "RPXX03"    "RPXX04"   
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{AU=name}}

This is used to select sequences having a specified author (only last name, no initial).

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("Graur", "AU=Graur")
 Graur$nelem
\end{Sinput}
\begin{Soutput}
[1] 48
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{AC=accession\_no}}

This is used to select sequences attached to specified accession number.
For instance if we are looking for sequences attached to the accession
number AY382159:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("ACexample", "AC=AY382159")
 getName(ACexample$req[[1]])
\end{Sinput}
\begin{Soutput}
[1] "AY382159"
\end{Soutput}
\begin{Sinput}
 annotations <- getAnnot(ACexample$req[[1]])
 cat(annotations, sep = "\n")
\end{Sinput}
\begin{Soutput}
ID   AY382159   standard; genomic DNA; PRO; 783 BP.
XX
AC   AY382159;
XX
SV   AY382159.1
XX
DT   08-OCT-2003 (Rel. 77, Created)
DT   08-OCT-2003 (Rel. 77, Last updated, Version 1)
XX
DE   Borrelia burgdorferi strain FP1 OspA gene, partial cds.
XX
KW   .
XX
OS   Borrelia burgdorferi (Lyme disease spirochete)
OC   Bacteria; Spirochaetes; Spirochaetales; Spirochaetaceae; Borrelia;
OC   Borrelia burgdorferi group.
XX
RN   [1]
RP   1-783
RA   Hao Q., Wan K.;
RT   ;
RL   Submitted (03-SEP-2003) to the EMBL/GenBank/DDBJ databases.
RL   Department of Lyme Spirochetosis, CDC, Beijing 102206, China
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..783
FT                   /db_xref="taxon:139"
FT                   /mol_type="genomic DNA"
FT                   /organism="Borrelia burgdorferi"
FT                   /strain="FP1"
FT   CDS             <1..>783
FT                   /codon_start=1
FT                   /transl_table=11
FT                   /product="OspA"
FT                   /protein_id="AAQ89576.1"
FT                   /translation="ALIACKQNVSSLDEKNSASVDLPGEMKVLVSKEKDKDGKYSLKAT
FT                   VDKLELKGTSDKNNGSGTLEGEKTDKSKAKLTISDDLSKTTFEVFKEDGKTLVSRKVSS
FT                   KDKTSTDEMFNEKGELSAKTMTRENGTKLEYTEMKSDGTGKTKEVLKNFTLEGRVANDK
FT                   VTLEVKEGTVTLSKEIAKSGEVTVALNDTNTTQATKKTGAWDSKTSTLTISVNSKKTTQ
FT                   LVFTKQDTITVQKYDSAGTNLEGTAVEIKTLDELKNALK"
XX
SQ   Sequence 783 BP; 342 A; 124 C; 145 G; 172 T; 0 other;
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{N=seq\_name}}

This is used to select sequences of a given name\footnote{
\textit{i.e.} what is documented in the ID or the LOCUS
field}. Sequences names are not necessarily stable, so that it's almost always
better to work with accession numbers. Anyway, the distinction between sequence names
and accession numbers is on a vanishing way because they tend more and
more to be the same thing (as in the example just below). 
The use of the at sign @ to substitute as a wildcard character for any zero or more characters 
is possible here.

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("Nexample", "N=AY382159")
 getName(Nexample$req[[1]])
\end{Sinput}
\begin{Soutput}
[1] "AY382159"
\end{Soutput}
\begin{Sinput}
 annotations <- getAnnot(Nexample$req[[1]])
 cat(annotations, sep = "\n")
\end{Sinput}
\begin{Soutput}
ID   AY382159   standard; genomic DNA; PRO; 783 BP.
XX
AC   AY382159;
XX
SV   AY382159.1
XX
DT   08-OCT-2003 (Rel. 77, Created)
DT   08-OCT-2003 (Rel. 77, Last updated, Version 1)
XX
DE   Borrelia burgdorferi strain FP1 OspA gene, partial cds.
XX
KW   .
XX
OS   Borrelia burgdorferi (Lyme disease spirochete)
OC   Bacteria; Spirochaetes; Spirochaetales; Spirochaetaceae; Borrelia;
OC   Borrelia burgdorferi group.
XX
RN   [1]
RP   1-783
RA   Hao Q., Wan K.;
RT   ;
RL   Submitted (03-SEP-2003) to the EMBL/GenBank/DDBJ databases.
RL   Department of Lyme Spirochetosis, CDC, Beijing 102206, China
XX
FH   Key             Location/Qualifiers
FH
FT   source          1..783
FT                   /db_xref="taxon:139"
FT                   /mol_type="genomic DNA"
FT                   /organism="Borrelia burgdorferi"
FT                   /strain="FP1"
FT   CDS             <1..>783
FT                   /codon_start=1
FT                   /transl_table=11
FT                   /product="OspA"
FT                   /protein_id="AAQ89576.1"
FT                   /translation="ALIACKQNVSSLDEKNSASVDLPGEMKVLVSKEKDKDGKYSLKAT
FT                   VDKLELKGTSDKNNGSGTLEGEKTDKSKAKLTISDDLSKTTFEVFKEDGKTLVSRKVSS
FT                   KDKTSTDEMFNEKGELSAKTMTRENGTKLEYTEMKSDGTGKTKEVLKNFTLEGRVANDK
FT                   VTLEVKEGTVTLSKEIAKSGEVTVALNDTNTTQATKKTGAWDSKTSTLTISVNSKKTTQ
FT                   LVFTKQDTITVQKYDSAGTNLEGTAVEIKTLDELKNALK"
XX
SQ   Sequence 783 BP; 342 A; 124 C; 145 G; 172 T; 0 other;
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{Y=year} or \texttt{Y>year} or \texttt{Y<year}}

This is used to select sequences published in a given year (\texttt{Y=year}),
or in a given year and after this year (\texttt{Y>year}), or in a given
year and before this year (\texttt{Y<year}).

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("Yexample", "Y=1999", virtual = TRUE)
 Yexample$nelem
\end{Sinput}
\begin{Soutput}
[1] 955274
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{O=organelle}}

This is used to select sequences from specified organelle named following defined code 
(\textit{e.g.}, chloroplast).
The list of available organelle codes can be obtained from the \texttt{readsmj()} function
this way:

\begin{Schunk}
\begin{Sinput}
 choosebank("genbank")
 nl <- readfirstrec(type = "SMJ")
 smj <- readsmj(nl = nl, all.add = TRUE)
 smj[!is.na(smj$nature) & smj$nature == "organelle", c("sname", 
     "libel")]
\end{Sinput}
\begin{Soutput}
             sname                    libel
3827   CHLOROPLAST       Chloroplast genome
3828 MITOCHONDRION     Mitochondrial genome
3829   NUCLEOMORPH       Nucleomorph genome
3830       PLASTID non-green plastid genome
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

To select for instance all sequences from chloroplast genome we can use:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("Oexample", "O=chloroplast", virtual = TRUE)
 Oexample$nelem
\end{Sinput}
\begin{Soutput}
[1] 65011
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{M=molecule}}

This is used to select sequences according to the chemical nature of the
sequenced molecule\footnote{as named in ID or LOCUS annotation records}.
The list of available organelle code can be obtained from the \texttt{readsmj()} function
this way:

\begin{Schunk}
\begin{Sinput}
 choosebank("genbank")
 nl <- readfirstrec(type = "SMJ")
 smj <- readsmj(nl = nl, all.add = TRUE)
 smj[!is.na(smj$nature) & smj$nature == "molecule", c("sname", 
     "libel")]
\end{Sinput}
\begin{Soutput}
    sname                                       libel
4    CRNA                                        <NA>
5     DNA                   Sequenced molecule is DNA
6    MRNA                  sequenced molecule is mRNA
7     RNA                   Sequenced molecule is RNA
8    RRNA                  sequenced molecule is rRNA
9   SCRNA sequenced molecule is small cytoplasmic RNA
10 SNORNA   sequenced molecule is small nucleolar RNA
11  SNRNA     sequenced molecule is small nuclear RNA
12   TRNA                  sequenced molecule is tRNA
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

To select for instance all sequences sequenced from DNA we can use:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("Mexample", "M=DNA", virtual = TRUE)
 Mexample$nelem
\end{Sinput}
\begin{Soutput}
[1] 7421752
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{ST=status}}

This is used to select sequences from specified data class (EMBL) or review level (UniProt).
The list of status codes can be obtained from the \texttt{readsmj()} function
this way:

\begin{Schunk}
\begin{Sinput}
 choosebank("embl")
 nl <- readfirstrec(type = "SMJ")
 smj <- readsmj(nl = nl, all.add = TRUE)
 smj[!is.na(smj$nature) & smj$nature == "status", c("sname", 
     "libel")]
\end{Sinput}
\begin{Soutput}
   sname                                        libel
1    ANN                     Annotated CON data class
2    EST           Expressed Sequence Tags data class
3    GSS            Genome Survey Sequence data class
4    HTC              High Throughput cDNA data class
5    HTG High Throughput Genome sequencing data class
6    PAT                            Patent data class
7    STD                          standard data class
8    STS              Sequence Tagged Site data class
9    TPA            Third Party Annotation data class
10   TSA    Transcriptome Shotgun Assembly data class
\end{Soutput}
\begin{Sinput}
 closebank()
 choosebank("swissprot")
 nl <- readfirstrec(type = "SMJ")
 smj <- readsmj(nl = nl, all.add = TRUE)
 smj[!is.na(smj$nature) & smj$nature == "status", c("sname", 
     "libel")]
\end{Sinput}
\begin{Soutput}
       sname                                                  libel
1   REVIEWED Entry was reviewed and annotated by UniProtKB curators
2 UNREVIEWED                               Computer-annotated entry
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

To select for instance all fully annotated sequences from Uniprot we can use:

\begin{Schunk}
\begin{Sinput}
 choosebank("swissprot")
 query("STexample", "ST=REVIEWED", virtual = TRUE)
 STexample$nelem
\end{Sinput}
\begin{Soutput}
[1] 392667
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{F=file\_name}}

This is used to select sequences whose names are in a given file, one name per line.
This is not directly implemented in seqinR, you have to use the function
\texttt{crelistfromclientdata()} or its short form \texttt{clfcd()} for this purpose. Here is an example with
a file of sequence names distributed with the seqinR package:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 fileSQ <- system.file("sequences/bb.mne", package = "seqinr")
 cat(readLines(fileSQ), sep = "\n")
\end{Sinput}
\begin{Soutput}
A04009.OSPA
A04009.OSPB
A22442
A24006
A24008
A24010
A24012
A24014
A24016
A33362
A67759.PE1
AB011063
AB011064
AB011065
AB011066
AB011067
AB035616
AB035617
AB035618
AB041949.VLSE
\end{Soutput}
\begin{Sinput}
 clfcd("listSQ", file = fileSQ, type = "SQ")
 getName(listSQ)
\end{Sinput}
\begin{Soutput}
 [1] "A04009.OSPA"   "A04009.OSPB"   "A22442"        "A24006"       
 [5] "A24008"        "A24010"        "A24012"        "A24014"       
 [9] "A24016"        "A33362"        "A67759.PE1"    "AB011063"     
[13] "AB011064"      "AB011065"      "AB011066"      "AB011067"     
[17] "AB035616"      "AB035617"      "AB035618"      "AB041949.VLSE"
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}


\subsection{\texttt{FA=file\_name}}

This is used to select sequences whose accession numbers are in a given file, one name per line.
This is not directly implemented in seqinR, you have to use the function
\texttt{crelistfromclientdata()} or its short form \texttt{clfcd()} for this purpose. Here is an example with
a file of sequence accession numbers distributed with the seqinR package:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 fileAC <- system.file("sequences/bb.acc", package = "seqinr")
 cat(readLines(fileAC), sep = "\n")
\end{Sinput}
\begin{Soutput}
AY382159
AY382160
AY491412
AY498719
AY498720
AY498721
AY498722
AY498723
AY498724
AY498725
AY498726
AY498727
AY498728
AY498729
AY499181
AY500379
AY500380
AY500381
AY500382
AY500383
\end{Soutput}
\begin{Sinput}
 clfcd("listAC", file = fileAC, type = "AC")
 getName(listAC)
\end{Sinput}
\begin{Soutput}
 [1] "AY382159" "AY382160" "AY491412" "AY498719" "AY498720" "AY498721"
 [7] "AY498722" "AY498723" "AY498724" "AY498725" "AY498726" "AY498727"
[13] "AY498728" "AY498729" "AY499181" "AY500379" "AY500380" "AY500381"
[19] "AY500382" "AY500383"
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}


\subsection{\texttt{FK=file\_name}}

This is used to produces the list of keywords named in given file, one keyword per line.
This is not directly implemented in seqinR, you have to use the function
\texttt{crelistfromclientdata()} or its short form \texttt{clfcd()} for this purpose. Here is an example with
a file of keywords distributed with the seqinR package:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 fileKW <- system.file("sequences/bb.kwd", package = "seqinr")
 cat(readLines(fileKW), sep = "\n")
\end{Sinput}
\begin{Soutput}
PLASMID
CIRCULAR
PARTIAL
5'-PARTIAL
3'-PARTIAL
MOTA GENE
MOTB GENE
DIVISION PRO
GYRB GENE
JOINING REGION
FTSA GENE
RPOB GENE
RPOC GENE
FLA GENE
DNAJ GENE
TUF GENE
PGK GENE
RUVA GENE
RUVB GENE
PROMOTER REGION
\end{Soutput}
\begin{Sinput}
 clfcd("listKW", file = fileKW, type = "KW")
 getName(listKW)
\end{Sinput}
\begin{Soutput}
 [1] "PLASMID"         "CIRCULAR"        "PARTIAL"         "5'-PARTIAL"     
 [5] "3'-PARTIAL"      "MOTA GENE"       "MOTB GENE"       "DIVISION PRO"   
 [9] "GYRB GENE"       "JOINING REGION"  "FTSA GENE"       "RPOB GENE"      
[13] "RPOC GENE"       "FLA GENE"        "DNAJ GENE"       "TUF GENE"       
[17] "PGK GENE"        "RUVA GENE"       "RUVB GENE"       "PROMOTER REGION"
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{FS=file\_name}}

This is used to produces the list of species named in given file, one species per line.
This is not directly implemented in seqinR, you have to use the function
\texttt{crelistfromclientdata()} or its short form \texttt{clfcd()} for this purpose. Here is an example with
a file of species names distributed with the seqinR package:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 fileSP <- system.file("sequences/bb.sp", package = "seqinr")
 cat(readLines(fileSP), sep = "\n")
\end{Sinput}
\begin{Soutput}
BORRELIA ANSERINA
BORRELIA CORIACEAE
BORRELIA PARKERI
BORRELIA TURICATAE
BORRELIA HERMSII
BORRELIA CROCIDURAE
BORRELIA LONESTARI
BORRELIA HISPANICA
BORRELIA BARBOURI
BORRELIA THEILERI
BORRELIA DUTTONII
BORRELIA MIYAMOTOI
BORRELIA PERSICA
BORRELIA RECURRENTIS
BORRELIA BURGDORFERI
BORRELIA AFZELII
BORRELIA GARINII
BORRELIA ANDERSONII
BORRELIA VALAISIANA
BORRELIA JAPONICA
\end{Soutput}
\begin{Sinput}
 clfcd("listSP", file = fileSP, type = "SP")
 getName(listSP)
\end{Sinput}
\begin{Soutput}
 [1] "BORRELIA ANSERINA"    "BORRELIA CORIACEAE"   "BORRELIA PARKERI"    
 [4] "BORRELIA TURICATAE"   "BORRELIA HERMSII"     "BORRELIA CROCIDURAE" 
 [7] "BORRELIA LONESTARI"   "BORRELIA HISPANICA"   "BORRELIA BARBOURI"   
[10] "BORRELIA THEILERI"    "BORRELIA DUTTONII"    "BORRELIA MIYAMOTOI"  
[13] "BORRELIA PERSICA"     "BORRELIA RECURRENTIS" "BORRELIA BURGDORFERI"
[16] "BORRELIA AFZELII"     "BORRELIA GARINII"     "BORRELIA ANDERSONII" 
[19] "BORRELIA VALAISIANA"  "BORRELIA JAPONICA"   
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{list\_name}}

A list name can be re-used, for instance:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("MyFirstListName", "Y=2000", virtual = TRUE)
 MyFirstListName$nelem
\end{Sinput}
\begin{Soutput}
[1] 885225
\end{Soutput}
\begin{Sinput}
 query("MySecondListName", "SP=Borrelia burgdorferi", virtual = TRUE)
 MySecondListName$nelem
\end{Sinput}
\begin{Soutput}
[1] 1682
\end{Soutput}
\begin{Sinput}
 query("MyThirdListName", "MyFirstListName AND MySecondListName", 
     virtual = TRUE)
 MyThirdListName$nelem
\end{Sinput}
\begin{Soutput}
[1] 131
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}
 
\section{Operators} 

\subsection{\texttt{AND}}

This is the binary operator for the logical and: a sequence belongs to the 
resulting list if, and only if, it is present in both operands. To select
for instance sequences from \textit{Borrelia burgdorferi} that are
also coding sequences we can use:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("ANDexample", "SP=Borrelia burgdorferi AND T=CDS", virtual = TRUE)
 ANDexample$nelem
\end{Sinput}
\begin{Soutput}
[1] 3218
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}


\subsection{\texttt{OR}}

This is the binary operator for the logical or: a sequence belongs to the 
resulting list if it is present in at least one of the two operands. To select
for instance sequences from \textit{Borrelia burgdorferi} or
from \textit{Escherichia coli} we can use:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("ORexample", "SP=Borrelia burgdorferi OR SP=Escherichia coli", 
     virtual = TRUE)
 ORexample$nelem
\end{Sinput}
\begin{Soutput}
[1] 28584
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{NOT}}

This is the unary operator for the logical negation. To select
for instance sequences from \textit{Borrelia burgdorferi} that
are not partial we can use:

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("NOTexample", "SP=Borrelia burgdorferi AND NOT K=PARTIAL", 
     virtual = TRUE)
 NOTexample$nelem
\end{Sinput}
\begin{Soutput}
[1] 3266
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{PAR}}

This is a unary operator to compute the list of parent sequences of a list
of sequences. The reciprocal operator is \texttt{SUB}. To check
the reciprocity we can use for instance:
 
\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("A", "T=TRNA", virtual = TRUE)
 query("B", "PAR A", virtual = TRUE)
 query("C", "SUB B", virtual = TRUE)
 query("D", "PAR C", virtual = TRUE)
 query("emptySet", "B AND NOT D", virtual = TRUE)
 emptySet$nelem
\end{Sinput}
\begin{Soutput}
[1] 0
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{SUB}}

This is a unary operator to add all subsequences of members of the single list operand.

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("SUBexample", "AC=AE000783", virtual = T)
 SUBexample$nelem
\end{Sinput}
\begin{Soutput}
[1] 70
\end{Soutput}
\begin{Sinput}
 query("SUBexample2", "SUB SUBexample", virtual = T)
 SUBexample2$nelem
\end{Sinput}
\begin{Soutput}
[1] 943
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{PS}}

This unary operator is used to get the list of species attached to member 
sequences of the operand list.

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("PSexample", "K=hyperthermo@", virtual = T)
 query("PSexample2", "PS PSexample")
 getName(PSexample2)
\end{Sinput}
\begin{Soutput}
[1] "BACILLUS LICHENIFORMIS" "DESULFUROCOCCUS"       
[3] "PYROCOCCUS FURIOSUS"   
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{PK}}

This unary operator is used to get the list of keywords attached to member sequences 
of the operand list.

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("PKexample", "AC=AE000783", virtual = T)
 query("PKexample2", "PK PKexample")
 getName(PKexample2)
\end{Sinput}
\begin{Soutput}
[1] "DIVISION PRO" "CDS"          "RRNA"         "TRNA"        
[5] "SOURCE"       "RELEASE 75"  
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{UN}}

This unary operator is used to get the list of sequences attached to a list
of species or keywords.

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 fileSP <- system.file("sequences/bb.sp", package = "seqinr")
 cat(readLines(fileSP), sep = "\n")
\end{Sinput}
\begin{Soutput}
BORRELIA ANSERINA
BORRELIA CORIACEAE
BORRELIA PARKERI
BORRELIA TURICATAE
BORRELIA HERMSII
BORRELIA CROCIDURAE
BORRELIA LONESTARI
BORRELIA HISPANICA
BORRELIA BARBOURI
BORRELIA THEILERI
BORRELIA DUTTONII
BORRELIA MIYAMOTOI
BORRELIA PERSICA
BORRELIA RECURRENTIS
BORRELIA BURGDORFERI
BORRELIA AFZELII
BORRELIA GARINII
BORRELIA ANDERSONII
BORRELIA VALAISIANA
BORRELIA JAPONICA
\end{Soutput}
\begin{Sinput}
 clfcd("listSP", file = fileSP, type = "SP")
 query("UNexample", "UN listSP", virtual = TRUE)
 UNexample$nelem
\end{Sinput}
\begin{Soutput}
[1] 2786
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{SD}}

This unary operator computes the list of species placed in the tree below the members of the 
species list operand.

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("hominidae", "SP=Hominidae", virtual = T)
 query("hsp", "PS hominidae", virtual = T)
 hsp$nelem
\end{Sinput}
\begin{Soutput}
[1] 19
\end{Soutput}
\begin{Sinput}
 query("SDexample", "SD hsp")
 getName(SDexample)
\end{Sinput}
\begin{Soutput}
 [1] "HOMINIDAE"                      "PONGO"                         
 [3] "PONGO PYGMAEUS"                 "PONGO PYGMAEUS ABELII"         
 [5] "PONGO PYGMAEUS PYGMAEUS"        "PONGO SP."                     
 [7] "HOMO/PAN/GORILLA GROUP"         "GORILLA"                       
 [9] "GORILLA GORILLA"                "GORILLA GORILLA BERINGEI"      
[11] "GORILLA GORILLA GRAUERI"        "GORILLA GORILLA GORILLA"       
[13] "GORILLA GORILLA UELLENSIS"      "PAN"                           
[15] "PAN TROGLODYTES"                "PAN TROGLODYTES SCHWEINFURTHII"
[17] "PAN TROGLODYTES TROGLODYTES"    "PAN TROGLODYTES VERUS"         
[19] "PAN TROGLODYTES VELLEROSUS"     "PAN PANISCUS"                  
[21] "HOMO"                           "HOMO SAPIENS"                  
[23] "HOMO SAPIENS NEANDERTHALENSIS" 
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}

\subsection{\texttt{KD}}

This unary operator computes the list of keywords placed in the tree below the members of the 
keywords list operand.

\begin{Schunk}
\begin{Sinput}
 choosebank("emblTP")
 query("cat", "SP=Felis catus", virtual = TRUE)
 query("catkw", "PK cat", virtual = TRUE)
 catkw$nelem
\end{Sinput}
\begin{Soutput}
[1] 540
\end{Soutput}
\begin{Sinput}
 query("KDexample", "KD catkw", virtual = TRUE)
 KDexample$nelem
\end{Sinput}
\begin{Soutput}
[1] 572
\end{Soutput}
\begin{Sinput}
 closebank()
\end{Sinput}
\end{Schunk}


\section*{Session Informations}

This part was compiled under the following \Rlogo{}~environment:

\begin{itemize}
  \item R version 2.8.0 (2008-10-20), \verb|i386-apple-darwin8.8.2|
  \item Locale: \verb|fr_FR.UTF-8/fr_FR.UTF-8/fr_FR.UTF-8/C/C/C|
  \item Base packages: base, datasets, grDevices, graphics, methods,
    stats, utils
  \item Other packages: MASS~7.2-44, ade4~1.4-9, ape~2.2-2,
    nlme~3.1-89, quadprog~1.4-11, seqinr~2.0-0, tseries~0.10-16,
    xtable~1.5-4, zoo~1.5-4
  \item Loaded via a namespace (and not attached): grid~2.8.0,
    lattice~0.17-15, tools~2.8.0
\end{itemize}
There were two compilation steps:

\begin{itemize}
  \item \Rlogo{} compilation time was: Sun Oct 26 18:04:21 2008
  \item \LaTeX{} compilation time was: \today
\end{itemize}

% END - DO NOT REMOVE THIS LINE

%%%%%%%%%%%%  BIBLIOGRAPHY %%%%%%%%%%%%%%%%%
\clearpage
\addcontentsline{toc}{section}{References}
\bibliographystyle{plain}
\bibliography{../config/book}
\end{document}
