\documentclass{article}
\input{../config/commontex}

\title{Test suite: run the don't run}
\author{Lobry, J.R.}

\usepackage{/Library/Frameworks/R.framework/Resources/share/texmf/Sweave}
\begin{document}
%
% To change the R input/output style:
%
\definecolor{Soutput}{rgb}{0,0,0.56}
\definecolor{Sinput}{rgb}{0.56,0,0}
\DefineVerbatimEnvironment{Sinput}{Verbatim}
{formatcom={\color{Sinput}},fontsize=\footnotesize, baselinestretch=0.75}
\DefineVerbatimEnvironment{Soutput}{Verbatim}
{formatcom={\color{Soutput}},fontsize=\footnotesize, baselinestretch=0.75}
%
% This removes the extra spacing after code and output chunks in Sweave,
% but keeps the spacing around the whole block.
%
\fvset{listparameters={\setlength{\topsep}{0pt}}}
\renewenvironment{Schunk}{\vspace{\topsep}}{\vspace{\topsep}}
%
% Rlogo
%
\newcommand{\Rlogo}{\protect\includegraphics[height=1.8ex,keepaspectratio]{../figs/Rlogo.pdf}}
%
% Shortcut for seqinR:
%
\newcommand{\seqinr}{\texttt{seqin\bf{R}}}
\newcommand{\Seqinr}{\texttt{Seqin\bf{R}}}
\fvset{fontsize= \scriptsize}
%
% R output options and libraries to be loaded.
%
%
%  Sweave Options
%
% Put all figures in the fig folder and start the name with current file name.
% Do not produce EPS files
%


\maketitle
\tableofcontents
% BEGIN - DO NOT REMOVE THIS LINE

\section{Introduction}

Many seqinR functions use socket connections to retrieve information from
the internet. As a consequence, most of examples should be protected by
a \verb!\dontrun{}! to pass the \texttt{R CMD CHECK}. In this section
we want to run automatically all these examples to check that everything
is OK.


\section{Stop list}

This is the list of function that don't run for now and need to be fixed.

\begin{Schunk}
\begin{Sinput}
 stoplist <- c("reverse.align", "extractseqs", "acnucopen", 
     "modifylist", "plot.SeqAcnucWeb", "draw.rearranged.oriloc")
\end{Sinput}
\end{Schunk}

Known problems are:

\begin{enumerate}
\item[reverse.align]{need clustalw on line, see later}
\item[extractseqs]{strange behaviour when in Sweave document???}
\item[acnucopen]{SUBINLNG was 60 and now 504}
\item[modifylist]{Error : mylist\$nelem == 33 is not TRUE}
\item[plot.SeqAcnucWeb]{Database with name -->hovernucl<-- is not known by server}
\item[draw.rearranged.oriloc]{Very long (infinite loop?)}
\end{enumerate}

\section{Figure list}

This is the list of functions that generates a graphical output.

\begin{Schunk}
\begin{Sinput}
 figlist <- c("draw.rearranged.oriloc", "oriloc", "dia.bactgensize", 
     "GC", "plot.SeqAcnucWeb")
\end{Sinput}
\end{Schunk}
\section{Don't run generator}

This code chunk generates the \texttt{dontrun.rnw} file that is included there after.
This file should be pre-existent, and two \texttt{Sweave()} passes are necessary.

\begin{Schunk}
\begin{Sinput}
 outfile <- file(paste(pwd, "dontrun.rnw", sep = "/"), open = "w")
 fex <- dir()
 for (f in fex) {
     fctname <- substr(x = f, start = 1, stop = nchar(f) - 
         2)
     if (fctname %in% stoplist) 
         next
     withfig <- "F"
     if (fctname %in% figlist) 
         withfig <- "T"
     lines <- readLines(f)
     dontrun <- lines[which(substring(lines, 1, 3) == "##D")]
     if (length(dontrun) == 0) 
         next
     dontrun <- sapply(dontrun, function(x) substr(x, 5, nchar(x)))
     writeLines(paste("\\subsection{\\texttt{", fctname, "()}}", 
         sep = ""), outfile)
     fctnamewithoutdots <- gsub("\\.", "", fctname)
     writeLines(paste("<<", fctnamewithoutdots, ",fig=", withfig, 
         ",keep.source=T>>=", sep = ""), outfile)
     writeLines(dontrun, outfile)
     writeLines("@", outfile)
 }
 close(outfile)
\end{Sinput}
\end{Schunk}


\begin{Schunk}
\begin{Sinput}
 setwd(pwd)
\end{Sinput}
\end{Schunk}

\subsection{\texttt{GC()}}
\begin{Schunk}
\begin{Sinput}
 # Too long for routine check
 # This is a benchmark to compare the effect of various parameter
 # setting on computation time
 n <- 10
 from <-10^4
 to <- 10^5
 size <- seq(from = from, to = to, length = n)
 res <- data.frame(matrix(NA, nrow = n, ncol = 5))
 colnames(res) <- c("size", "FF", "FT", "TF", "TT")
 res[, "size"] <- size
 for(i in seq_len(n)){
   myseq <- sample(x = s2c("acgtws"), size = size[i], replace = TRUE)   
   res[i, "FF"] <- system.time(GC(myseq, forceToLower = FALSE, exact = FALSE))[3]
   res[i, "FT"] <- system.time(GC(myseq, forceToLower = FALSE, exact = TRUE))[3]
         res[i, "TF"] <- system.time(GC(myseq, forceToLower = TRUE, exact = FALSE))[3]
         res[i, "TT"] <- system.time(GC(myseq, forceToLower = TRUE, exact = TRUE))[3]
 }
 par(oma = c(0,0,2.5,0), mar = c(4,5,0,2) + 0.1, mfrow = c(2, 1))
 plot(res$size, res$TT, las = 1, 
 xlab = "Sequence size [bp]",
 ylim = c(0, max(res$TT)), xlim = c(0, max(res$size)), ylab = "")
 title(ylab = "Observed time [s]", line = 4)
 abline(lm(res$TT~res$size))
 points(res$size, res$FT, col = "red")
 abline(lm(res$FT~res$size), col = "red", lty = 3)
 points(res$size, res$TF, pch = 2)
 abline(lm(res$TF~res$size))
 points(res$size, res$FF, pch = 2, col = "red")
 abline(lm(res$FF~res$size), lty = 3, col = "red")
 legend("topleft", inset = 0.01, legend = c("forceToLower = TRUE", "forceToLower = FALSE"), col = c("black", "red"), lty = c(1,3))
 legend("bottomright", inset = 0.01, legend = c("exact = TRUE", "exact = FALSE"),
 pch = c(1,2))
 mincpu <- lm(res$FF~res$size)$coef[2]
 barplot(
 c(lm(res$FF~res$size)$coef[2]/mincpu, 
   lm(res$TF~res$size)$coef[2]/mincpu,
   lm(res$FT~res$size)$coef[2]/mincpu,
   lm(res$TT~res$size)$coef[2]/mincpu),
 horiz = TRUE, xlab = "Increase of CPU time",
 col = c("red", "black", "red", "black"),
 names.arg = c("(F,F)", "(T,F)", "(F,T)", "(T,T)"), las = 1)
 title(ylab = "forceToLower,exact", line = 4)
 mtext("CPU time as function of options", outer = TRUE, line = 1, cex = 1.5)
\end{Sinput}
\end{Schunk}
\includegraphics{../figs/rundontrun-GC}
\subsection{\texttt{SeqAcnucWeb()}}
\begin{Schunk}
\begin{Sinput}
 # Need internet connection
   choosebank("emblTP")
   query("mylist", "sp=felis catus et t=cds et o=mitochondrion")
   stopifnot(is.SeqAcnucWeb(mylist$req[[1]]))
   closebank()
\end{Sinput}
\end{Schunk}
\subsection{\texttt{alllistranks()}}
\begin{Schunk}
\begin{Sinput}
 # Need internet connection
  choosebank("emblTP")
  query("tmp1", "sp=Borrelia burgdorferi", virtual = TRUE)
  query("tmp2", "sp=Borrelia burgdorferi", virtual = TRUE)
  query("tmp3", "sp=Borrelia burgdorferi", virtual = TRUE)
  (result <- alllistranks())
\end{Sinput}
\begin{Soutput}
$count
[1] 3

$ranks
[1] 2 3 4
\end{Soutput}
\begin{Sinput}
  stopifnot(result$count == 3)   # Three ACNUC lists
  stopifnot(result$ranks == 2:4) # Starting at rank 2
  #
  # Summay of current lists defined on the ACNUC server:
  #
  sapply(result$ranks, getliststate)
\end{Sinput}
\begin{Soutput}
      [,1]   [,2]   [,3]  
type  "SQ"   "SQ"   "SQ"  
name  "TMP1" "TMP2" "TMP3"
count 1682   1682   1682  
locus TRUE   TRUE   TRUE  
\end{Soutput}
\begin{Sinput}
  closebank()
\end{Sinput}
\end{Schunk}
\subsection{\texttt{autosocket()}}
\begin{Schunk}
\begin{Sinput}
  #Need internet connection
   choosebank("emblTP")
   autosocket()
\end{Sinput}
\begin{Soutput}
                description                       class 
"->pbil.univ-lyon1.fr:5558"                  "sockconn" 
                       mode                        text 
                       "a+"                      "text" 
                     opened                    can read 
                   "opened"                       "yes" 
                  can write 
                      "yes" 
\end{Soutput}
\begin{Sinput}
   closebank()
\end{Sinput}
\end{Schunk}
\subsection{\texttt{choosebank()}}
\begin{Schunk}
\begin{Sinput}
 # Need internet connection
   # Show available databases:  
   choosebank()
\end{Sinput}
\begin{Soutput}
 [1] "genbank"     "embl"        "emblwgs"     "swissprot"   "ensembl"    
 [6] "refseq"      "nrsub"       "hobacnucl"   "hobacprot"   "hovergendna"
[11] "hovergen"    "hogenom"     "hogenomdna"  "hogennucl"   "hogenprot"  
[16] "hoverclnu"   "hoverclpr"   "homolens"    "homolensdna" "greview"    
[21] "polymorphix" "emglib"      "HAMAPnucl"   "HAMAPprot"   "hoppsigen"  
[26] "nurebnucl"   "nurebprot"   "taxobacgen" 
\end{Soutput}
\begin{Sinput}
   # Show frozen databases:
   choosebank(tag = "TP")
\end{Sinput}
\begin{Soutput}
[1] "emblTP"      "swissprotTP" "hoverprotTP" "hovernuclTP" "trypano"    
\end{Soutput}
\begin{Sinput}
   # Select a database:
   choosebank("emblTP", tag = "TP") 
   # Do something with the database:
   myseq <- gfrag("LMFLCHR36", start = 1, length = 30)
   stopifnot(myseq == "cgcgtgctggcggcaatgaagcgttcgatg")
   # Close the database:
   closebank()
\end{Sinput}
\end{Schunk}
\subsection{\texttt{closebank()}}
\begin{Schunk}
\begin{Sinput}
 # Need internet connection
    choosebank("emblTP")
    closebank()
\end{Sinput}
\end{Schunk}
\subsection{\texttt{countfreelists()}}
\begin{Schunk}
\begin{Sinput}
  # Need internet connection
   choosebank("emblTP")
   (rescountfreelists <- countfreelists())
\end{Sinput}
\begin{Soutput}
$free
[1] 48

$annotlines
 [1] "ALL" "AC"  "PR"  "DT"  "KW"  "OS"  "OC"  "OG"  "RN"  "RC"  "RP"  "RX" 
[13] "RG"  "RA"  "RT"  "RL"  "DR"  "CC"  "AH"  "AS"  "FH"  "FT"  "CO"  "SQ" 
[25] "SEQ"
\end{Soutput}
\begin{Sinput}
   stopifnot(all(rescountfreelists$annotlines == 
    c("ALL", "AC",  "PR",  "DT",  "KW",  "OS",  "OC",
    "OG",  "RN",  "RC",  "RP",  "RX", "RG",  "RA",  "RT",  "RL",  "DR", 
    "CC",  "AH",  "AS",  "FH",  "FT",  "CO",  "SQ", "SEQ")))
   closebank()
\end{Sinput}
\end{Schunk}
\subsection{\texttt{countsubseqs()}}
\begin{Schunk}
\begin{Sinput}
  # Need internet connection
   choosebank("emblTP")
   query("mylist", "N=@", virtual = TRUE) # select all (seqs + subseqs)
   mylist$nelem   # 14138094 seqs + subseqs
\end{Sinput}
\begin{Soutput}
[1] 14138094
\end{Soutput}
\begin{Sinput}
   stopifnot(mylist$nelem == 14138094)
   css(glr("mylist")) # 1604500 subsequences only
\end{Sinput}
\begin{Soutput}
[1] 1604500
\end{Soutput}
\begin{Sinput}
   stopifnot(css(glr("mylist")) == 1604500)
   closebank()
\end{Sinput}
\end{Schunk}
\subsection{\texttt{crelistfromclientdata()}}
\begin{Schunk}
\begin{Sinput}
  # Need internet connection
  choosebank("emblTP")
  #
  # Example with a file that contains sequence names:
  #
  fileSQ <- system.file("sequences/bb.mne", package = "seqinr")
  crelistfromclientdata("listSQ", file = fileSQ, type = "SQ")
  sapply(listSQ$req, getName)
\end{Sinput}
\begin{Soutput}
 [1] "A04009.OSPA"   "A04009.OSPB"   "A22442"        "A24006"       
 [5] "A24008"        "A24010"        "A24012"        "A24014"       
 [9] "A24016"        "A33362"        "A67759.PE1"    "AB011063"     
[13] "AB011064"      "AB011065"      "AB011066"      "AB011067"     
[17] "AB035616"      "AB035617"      "AB035618"      "AB041949.VLSE"
\end{Soutput}
\begin{Sinput}
  #
  # Example with a file that contains sequence accession numbers:
  #
  fileAC <- system.file("sequences/bb.acc", package = "seqinr")
  crelistfromclientdata("listAC", file = fileAC, type = "AC")
  sapply(listAC$req, getName) 
\end{Sinput}
\begin{Soutput}
 [1] "AY382159" "AY382160" "AY491412" "AY498719" "AY498720" "AY498721"
 [7] "AY498722" "AY498723" "AY498724" "AY498725" "AY498726" "AY498727"
[13] "AY498728" "AY498729" "AY499181" "AY500379" "AY500380" "AY500381"
[19] "AY500382" "AY500383"
\end{Soutput}
\begin{Sinput}
  #
  # Example with a file that contains species names:
  #
  fileSP <- system.file("sequences/bb.sp", package = "seqinr")
  crelistfromclientdata("listSP", file = fileSP, type = "SP")
  sapply(listSP$req, getName) 
\end{Sinput}
\begin{Soutput}
 [1] "BORRELIA ANSERINA"    "BORRELIA CORIACEAE"   "BORRELIA PARKERI"    
 [4] "BORRELIA TURICATAE"   "BORRELIA HERMSII"     "BORRELIA CROCIDURAE" 
 [7] "BORRELIA LONESTARI"   "BORRELIA HISPANICA"   "BORRELIA BARBOURI"   
[10] "BORRELIA THEILERI"    "BORRELIA DUTTONII"    "BORRELIA MIYAMOTOI"  
[13] "BORRELIA PERSICA"     "BORRELIA RECURRENTIS" "BORRELIA BURGDORFERI"
[16] "BORRELIA AFZELII"     "BORRELIA GARINII"     "BORRELIA ANDERSONII" 
[19] "BORRELIA VALAISIANA"  "BORRELIA JAPONICA"   
\end{Soutput}
\begin{Sinput}
  #
  # Example with a file that contains keywords:
  #
  fileKW <- system.file("sequences/bb.kwd", package = "seqinr")
  crelistfromclientdata("listKW", file = fileKW, type = "KW")
  sapply(listKW$req, getName)
\end{Sinput}
\begin{Soutput}
 [1] "PLASMID"         "CIRCULAR"        "PARTIAL"         "5'-PARTIAL"     
 [5] "3'-PARTIAL"      "MOTA GENE"       "MOTB GENE"       "DIVISION PRO"   
 [9] "GYRB GENE"       "JOINING REGION"  "FTSA GENE"       "RPOB GENE"      
[13] "RPOC GENE"       "FLA GENE"        "DNAJ GENE"       "TUF GENE"       
[17] "PGK GENE"        "RUVA GENE"       "RUVB GENE"       "PROMOTER REGION"
\end{Soutput}
\begin{Sinput}
  #
  # Summary of ACNUC lists:
  #
  sapply(alr()$rank, getliststate)
\end{Sinput}
\begin{Soutput}
      [,1]     [,2]     [,3]     [,4]    
type  "SQ"     "SQ"     "SP"     "KW"    
name  "LISTSQ" "LISTAC" "LISTSP" "LISTKW"
count 20       20       20       20      
locus FALSE    TRUE     TRUE     TRUE    
\end{Soutput}
\begin{Sinput}
  closebank() 
\end{Sinput}
\end{Schunk}
\subsection{\texttt{dia.bactgensize()}}
\begin{Schunk}
\begin{Sinput}
   # Need internet connection
   dia.bactgensize(source = "http://www.genomesonline.org/DBs/goldtable.txt")
\end{Sinput}
\end{Schunk}
\includegraphics{../figs/rundontrun-diabactgensize}
\subsection{\texttt{extract.breakpoints()}}
\begin{Schunk}
\begin{Sinput}
 r.ori <- rearranged.oriloc(seq.fasta = system.file("sequences/ct.fasta",package = "seqinr"),
     g2.coord = system.file("sequences/ct.coord",package = "seqinr"))
\end{Sinput}
\end{Schunk}
\subsection{\texttt{getAnnot()}}
\begin{Schunk}
\begin{Sinput}
   # Need internet connection
   choosebank("emblTP")
   query("fc", "sp=felis catus et t=cds et O=mitochondrion et Y>2001 et no k=partial")
   # get the first 5 lines annotating the first sequence:
   annots <- getAnnot(fc$req[[1]], nbl = 5)
   cat(annots, sep = "\n")
\end{Sinput}
\begin{Soutput}
FT   CDS             100..303
FT                   /db_xref="GOA:Q94NW9"
FT                   /db_xref="TrEMBL:Q94NW9"
FT                   /transl_table=2
FT                   /gene="ATPase8"
\end{Soutput}
\begin{Sinput}
   # or use the list method to get them all at once:
   annots <- getAnnot(fc$req, nbl = 5)
   cat(annots, sep = "\n")
\end{Sinput}
\begin{Soutput}
FT   CDS             100..303
FT                   /db_xref="GOA:Q94NW9"
FT                   /db_xref="TrEMBL:Q94NW9"
FT                   /transl_table=2
FT                   /gene="ATPase8"
FT   CDS             100..303
FT                   /db_xref="GOA:Q94NW9"
FT                   /db_xref="TrEMBL:Q94NW9"
FT                   /transl_table=2
FT                   /gene="ATPase8"
FT   CDS             100..303
FT                   /db_xref="GOA:Q94NW9"
FT                   /db_xref="TrEMBL:Q94NW9"
FT                   /transl_table=2
FT                   /gene="ATPase8"
FT   CDS             100..303
FT                   /db_xref="GOA:Q94NW9"
FT                   /db_xref="TrEMBL:Q94NW9"
FT                   /transl_table=2
FT                   /gene="ATPase8"
FT   CDS             100..303
FT                   /db_xref="GOA:Q94NW9"
FT                   /db_xref="TrEMBL:Q94NW9"
FT                   /transl_table=2
FT                   /gene="ATPase8"
FT   CDS             100..303
FT                   /db_xref="GOA:Q94NW9"
FT                   /db_xref="TrEMBL:Q94NW9"
FT                   /transl_table=2
FT                   /gene="ATPase8"
FT   CDS             100..303
FT                   /db_xref="GOA:Q94NW9"
FT                   /db_xref="TrEMBL:Q94NW9"
FT                   /transl_table=2
FT                   /gene="ATPase8"
\end{Soutput}
\begin{Sinput}
   closebank()
\end{Sinput}
\end{Schunk}
\subsection{\texttt{getKeyword()}}
\begin{Schunk}
\begin{Sinput}
   # Need internet connection
   choosebank("emblTP")
   query("fc", "sp=felis catus et t=cds et o=mitochondrion")
   getKeyword(fc$req[[1]])
\end{Sinput}
\begin{Soutput}
[1] "DIVISION ORG" "RELEASE 62"   "CYTOCHROME B" "SOURCE"      
[5] "CDS"         
\end{Soutput}
\begin{Sinput}
   # Should be: 
   # [1] "DIVISION ORG" "RELEASE 62"   "CYTOCHROME B" "SOURCE"       "CDS"
   closebank()  
\end{Sinput}
\end{Schunk}
\subsection{\texttt{getLength()}}
\begin{Schunk}
\begin{Sinput}
   # Need internet connection
   choosebank("emblTP")
   query("fc", "sp=felis catus et t=cds et o=mitochondrion")
   getLength(fc)
\end{Sinput}
\begin{Soutput}
 [1] 1140 1140  300  402  402   27  204   95   27  204   95   27  204   95
[15]   27  204   95   27  204   95   27  204   95   27  204   95  316  402
[29]  957 1042 1545  684  204  681  784  347  297 1378 1821  528 1140 1140
[43]  345  319  237  250  345  237  249
\end{Soutput}
\begin{Sinput}
   closebank()  
\end{Sinput}
\end{Schunk}
\subsection{\texttt{getLocation()}}
\begin{Schunk}
\begin{Sinput}
   # Need internet connection
   choosebank("emblTP")
   query("fc", "sp=felis catus et t=cds et o=mitochondrion")
   getLocation(fc$req[[5]])
\end{Sinput}
\begin{Soutput}
[1]  27 428
\end{Soutput}
\begin{Sinput}
   closebank()  
\end{Sinput}
\end{Schunk}
\subsection{\texttt{getName()}}
\begin{Schunk}
\begin{Sinput}
   # Need internet connection
   choosebank("emblTP")
   query("fc", "sp=felis catus et t=cds et o=mitochondrion")
   getName(fc)
\end{Sinput}
\begin{Soutput}
 [1] "AB004237"        "AB004238"        "AF172359"        "FCA300702"      
 [5] "FCA441328.CYTB"  "FSI409128.COII"  "FSI409128.PE2"   "FSI409128.PE3"  
 [9] "FSI409129.COII"  "FSI409129.PE2"   "FSI409129.PE3"   "FSI409130.COII" 
[13] "FSI409130.PE2"   "FSI409130.PE3"   "FSI409131.COII"  "FSI409131.PE2"  
[17] "FSI409131.PE3"   "FSI409132.COII"  "FSI409132.PE2"   "FSI409132.PE3"  
[21] "FSI409133.COII"  "FSI409133.PE2"   "FSI409133.PE3"   "FSI409134.COII" 
[25] "FSI409134.PE2"   "FSI409134.PE3"   "MI1290634.PE1"   "MIFCCBD"        
[29] "MIFCCU207.ND1"   "MIFCCU207.ND2"   "MIFCCU207.COI"   "MIFCCU207.COII" 
[33] "MIFCCU207.PE5"   "MIFCCU207.PE6"   "MIFCCU207.COIII" "MIFCCU207.ND3"  
[37] "MIFCCU207.ND4L"  "MIFCCU207.ND4"   "MIFCCU207.ND5"   "MIFCCU207.ND6"  
[41] "MIFCCU207.CYTB"  "MIFDCYTB"        "S75096"          "S75098"         
[45] "S75099.COI"      "S75101"          "S75328"          "S75331.COII"    
[49] "S75332.COI"     
\end{Soutput}
\begin{Sinput}
   closebank()  
\end{Sinput}
\end{Schunk}
\subsection{\texttt{getSequence()}}
\begin{Schunk}
\begin{Sinput}
 # Need internet connection
   choosebank("emblTP")
   query("fc", "sp=felis catus et t=cds et o=mitochondrion")
   getSequence(fc$req[[1]])
\end{Sinput}
\begin{Soutput}
   [1] "a" "t" "g" "a" "c" "c" "a" "a" "c" "a" "t" "t" "c" "g" "a" "a" "a"
  [18] "a" "t" "c" "a" "c" "a" "c" "c" "c" "c" "c" "t" "t" "a" "c" "c" "a"
  [35] "a" "a" "a" "t" "t" "a" "t" "t" "a" "a" "t" "c" "a" "c" "t" "c" "a"
  [52] "t" "t" "c" "a" "t" "c" "g" "a" "c" "c" "t" "a" "c" "c" "t" "g" "c"
  [69] "c" "c" "c" "a" "t" "c" "t" "a" "a" "c" "a" "t" "c" "t" "c" "a" "g"
  [86] "c" "a" "t" "g" "a" "t" "g" "a" "a" "a" "c" "t" "t" "c" "g" "g" "c"
 [103] "t" "c" "c" "c" "t" "t" "c" "t" "a" "g" "g" "a" "g" "t" "c" "t" "g"
 [120] "c" "c" "t" "a" "a" "t" "c" "t" "t" "a" "c" "a" "a" "a" "t" "c" "c"
 [137] "t" "c" "a" "c" "c" "g" "g" "c" "c" "t" "c" "t" "t" "t" "t" "t" "g"
 [154] "g" "c" "c" "a" "t" "a" "c" "a" "c" "t" "a" "c" "a" "c" "a" "t" "c"
 [171] "a" "g" "a" "c" "a" "c" "a" "a" "c" "a" "a" "c" "c" "g" "c" "c" "t"
 [188] "t" "t" "t" "c" "a" "t" "c" "a" "g" "t" "t" "a" "c" "c" "c" "a" "c"
 [205] "a" "t" "c" "t" "g" "t" "c" "g" "c" "g" "a" "c" "g" "t" "t" "a" "a"
 [222] "t" "t" "a" "t" "g" "g" "c" "t" "g" "a" "a" "t" "c" "a" "t" "c" "c"
 [239] "g" "a" "t" "a" "t" "t" "t" "a" "c" "a" "c" "g" "c" "c" "a" "a" "c"
 [256] "g" "g" "a" "g" "c" "t" "t" "c" "t" "a" "t" "a" "t" "t" "c" "t" "t"
 [273] "t" "a" "t" "c" "t" "g" "c" "c" "t" "g" "t" "a" "c" "a" "t" "a" "c"
 [290] "a" "t" "g" "t" "a" "g" "g" "a" "c" "g" "g" "g" "g" "a" "a" "t" "a"
 [307] "t" "a" "c" "t" "a" "c" "g" "g" "c" "t" "c" "c" "t" "a" "c" "a" "c"
 [324] "c" "t" "t" "c" "t" "c" "a" "g" "a" "g" "a" "c" "a" "t" "g" "a" "a"
 [341] "a" "c" "a" "t" "t" "g" "g" "a" "a" "t" "c" "a" "t" "a" "c" "t" "a"
 [358] "t" "t" "a" "t" "t" "t" "a" "c" "a" "g" "t" "c" "a" "t" "a" "g" "c"
 [375] "c" "a" "c" "a" "g" "c" "t" "t" "t" "t" "a" "t" "g" "g" "g" "a" "t"
 [392] "a" "c" "g" "t" "c" "c" "t" "a" "c" "c" "a" "t" "g" "a" "g" "g" "c"
 [409] "c" "a" "a" "a" "t" "g" "t" "c" "c" "t" "t" "c" "t" "g" "a" "g" "g"
 [426] "a" "g" "c" "a" "a" "c" "c" "g" "t" "a" "a" "t" "c" "a" "c" "t" "a"
 [443] "a" "c" "c" "t" "c" "c" "t" "g" "t" "c" "a" "g" "c" "a" "a" "t" "t"
 [460] "c" "c" "a" "t" "a" "c" "a" "t" "c" "g" "g" "g" "a" "c" "t" "g" "a"
 [477] "a" "c" "t" "a" "g" "t" "a" "g" "a" "a" "t" "g" "a" "a" "t" "c" "t"
 [494] "g" "a" "g" "g" "g" "g" "g" "c" "t" "t" "c" "t" "c" "a" "g" "t" "a"
 [511] "g" "a" "c" "a" "a" "a" "g" "c" "c" "a" "c" "c" "c" "t" "a" "a" "c"
 [528] "a" "c" "g" "a" "t" "t" "c" "t" "t" "t" "g" "c" "t" "t" "t" "c" "c"
 [545] "a" "c" "t" "t" "c" "a" "t" "t" "c" "t" "t" "c" "c" "a" "t" "t" "c"
 [562] "a" "t" "t" "a" "t" "c" "t" "c" "a" "g" "c" "c" "t" "t" "a" "g" "c"
 [579] "a" "g" "c" "a" "g" "t" "a" "c" "a" "c" "c" "t" "c" "t" "t" "a" "t"
 [596] "t" "c" "c" "t" "t" "c" "a" "t" "g" "a" "a" "a" "c" "a" "g" "g" "a"
 [613] "t" "c" "t" "a" "a" "c" "a" "a" "c" "c" "c" "c" "t" "c" "a" "g" "g"
 [630] "a" "a" "t" "t" "a" "c" "a" "t" "c" "c" "g" "a" "t" "t" "c" "a" "g"
 [647] "a" "c" "a" "a" "a" "a" "t" "c" "c" "c" "a" "t" "t" "c" "c" "a" "c"
 [664] "c" "c" "a" "t" "a" "c" "t" "a" "t" "a" "c" "a" "a" "t" "c" "a" "a"
 [681] "a" "g" "a" "c" "a" "t" "c" "c" "t" "a" "g" "g" "t" "c" "t" "t" "c"
 [698] "t" "a" "g" "t" "a" "c" "t" "a" "g" "t" "t" "t" "t" "a" "a" "c" "a"
 [715] "c" "t" "c" "a" "t" "a" "c" "t" "a" "c" "t" "c" "g" "t" "c" "c" "t"
 [732] "a" "t" "t" "t" "t" "c" "a" "c" "c" "a" "g" "a" "c" "c" "t" "g" "c"
 [749] "t" "a" "g" "g" "a" "g" "a" "c" "c" "c" "a" "g" "a" "c" "a" "a" "c"
 [766] "t" "a" "c" "a" "t" "c" "c" "c" "a" "g" "c" "c" "a" "a" "c" "c" "c"
 [783] "t" "t" "t" "a" "a" "a" "t" "a" "c" "c" "c" "c" "t" "c" "c" "c" "c"
 [800] "a" "t" "a" "t" "t" "a" "a" "a" "c" "c" "t" "g" "a" "a" "t" "g" "a"
 [817] "t" "a" "c" "t" "t" "c" "c" "t" "a" "t" "t" "c" "g" "c" "a" "t" "a"
 [834] "c" "g" "c" "a" "a" "t" "t" "c" "t" "c" "c" "g" "a" "t" "c" "c" "a"
 [851] "t" "c" "c" "c" "c" "a" "a" "c" "a" "a" "a" "c" "t" "a" "g" "g" "g"
 [868] "g" "g" "a" "g" "t" "c" "c" "t" "a" "g" "c" "c" "c" "t" "a" "g" "t"
 [885] "a" "c" "t" "c" "t" "c" "c" "a" "t" "c" "c" "t" "a" "g" "t" "a" "c"
 [902] "t" "a" "g" "c" "a" "a" "t" "c" "a" "t" "t" "c" "c" "a" "a" "t" "c"
 [919] "c" "t" "c" "c" "a" "c" "a" "c" "c" "t" "c" "c" "a" "a" "a" "c" "a"
 [936] "a" "c" "g" "a" "g" "g" "a" "a" "t" "a" "a" "t" "g" "t" "t" "t" "c"
 [953] "g" "a" "c" "c" "a" "c" "t" "a" "a" "g" "c" "c" "a" "a" "t" "g" "t"
 [970] "c" "t" "a" "t" "t" "c" "t" "g" "a" "c" "t" "c" "c" "t" "a" "g" "t"
 [987] "a" "g" "c" "g" "g" "a" "t" "c" "t" "c" "c" "t" "a" "a" "c" "c" "c"
[1004] "t" "a" "a" "c" "a" "t" "g" "a" "a" "t" "c" "g" "g" "t" "g" "g" "c"
[1021] "c" "a" "a" "c" "c" "t" "g" "t" "a" "g" "a" "a" "c" "a" "t" "c" "c"
[1038] "a" "t" "t" "c" "a" "t" "c" "a" "c" "c" "a" "t" "c" "g" "g" "c" "c"
[1055] "a" "a" "c" "t" "a" "g" "c" "c" "t" "c" "c" "a" "t" "c" "c" "t" "a"
[1072] "t" "a" "t" "t" "t" "c" "t" "c" "a" "a" "c" "c" "c" "t" "c" "c" "t"
[1089] "a" "a" "t" "c" "c" "t" "a" "a" "t" "a" "c" "c" "c" "a" "t" "c" "t"
[1106] "c" "a" "g" "g" "c" "a" "t" "t" "a" "t" "t" "g" "a" "a" "a" "a" "c"
[1123] "c" "g" "c" "c" "t" "a" "c" "t" "c" "a" "a" "a" "t" "g" "a" "a" "g"
[1140] "a"
\end{Soutput}
\begin{Sinput}
   getSequence(fc$req[[1]], as.string = TRUE)
\end{Sinput}
\begin{Soutput}
[1] "atgaccaacattcgaaaatcacacccccttaccaaaattattaatcactcattcatcgacctacctgccccatctaacatctcagcatgatgaaacttcggctcccttctaggagtctgcctaatcttacaaatcctcaccggcctctttttggccatacactacacatcagacacaacaaccgccttttcatcagttacccacatctgtcgcgacgttaattatggctgaatcatccgatatttacacgccaacggagcttctatattctttatctgcctgtacatacatgtaggacggggaatatactacggctcctacaccttctcagagacatgaaacattggaatcatactattatttacagtcatagccacagcttttatgggatacgtcctaccatgaggccaaatgtccttctgaggagcaaccgtaatcactaacctcctgtcagcaattccatacatcgggactgaactagtagaatgaatctgagggggcttctcagtagacaaagccaccctaacacgattctttgctttccacttcattcttccattcattatctcagccttagcagcagtacacctcttattccttcatgaaacaggatctaacaacccctcaggaattacatccgattcagacaaaatcccattccacccatactatacaatcaaagacatcctaggtcttctagtactagttttaacactcatactactcgtcctattttcaccagacctgctaggagacccagacaactacatcccagccaaccctttaaatacccctccccatattaaacctgaatgatacttcctattcgcatacgcaattctccgatccatccccaacaaactagggggagtcctagccctagtactctccatcctagtactagcaatcattccaatcctccacacctccaaacaacgaggaataatgtttcgaccactaagccaatgtctattctgactcctagtagcggatctcctaaccctaacatgaatcggtggccaacctgtagaacatccattcatcaccatcggccaactagcctccatcctatatttctcaaccctcctaatcctaatacccatctcaggcattattgaaaaccgcctactcaaatgaaga"
\end{Soutput}
\begin{Sinput}
   closebank()
\end{Sinput}
\end{Schunk}
\subsection{\texttt{getTrans()}}
\begin{Schunk}
\begin{Sinput}
   # Need internet connection.
   # Translation of the following EMBL entry:
   #
   # FT   CDS             join(complement(153944..154157),complement(153727..153866),
   # FT                   complement(152185..153037),138523..138735,138795..138955)
   # FT                   /codon_start=1
   choosebank("emblTP")
   query("trans", "N=AE003734.PE35")
   getTrans(trans$req[[1]])
\end{Sinput}
\begin{Soutput}
  [1] "M" "A" "D" "D" "E" "Q" "F" "S" "L" "C" "W" "N" "N" "F" "N" "T" "N" "L"
 [19] "S" "A" "G" "F" "H" "E" "S" "L" "C" "R" "G" "D" "L" "V" "D" "V" "S" "L"
 [37] "A" "A" "E" "G" "Q" "I" "V" "K" "A" "H" "R" "L" "V" "L" "S" "V" "C" "S"
 [55] "P" "F" "F" "R" "K" "M" "F" "T" "Q" "M" "P" "S" "N" "T" "H" "A" "I" "V"
 [73] "F" "L" "N" "N" "V" "S" "H" "S" "A" "L" "K" "D" "L" "I" "Q" "F" "M" "Y"
 [91] "C" "G" "E" "V" "N" "V" "K" "Q" "D" "A" "L" "P" "A" "F" "I" "S" "T" "A"
[109] "E" "S" "L" "Q" "I" "K" "G" "L" "T" "D" "N" "D" "P" "A" "P" "Q" "P" "P"
[127] "Q" "E" "S" "S" "P" "P" "P" "A" "A" "P" "H" "V" "Q" "Q" "Q" "Q" "I" "P"
[145] "A" "Q" "R" "V" "Q" "R" "Q" "Q" "P" "R" "A" "S" "A" "R" "Y" "K" "I" "E"
[163] "T" "V" "D" "D" "G" "L" "G" "D" "E" "K" "Q" "S" "T" "T" "Q" "I" "V" "I"
[181] "Q" "T" "T" "A" "A" "P" "Q" "A" "T" "I" "V" "Q" "Q" "Q" "Q" "P" "Q" "Q"
[199] "A" "A" "Q" "Q" "I" "Q" "S" "Q" "Q" "L" "Q" "T" "G" "T" "T" "T" "T" "A"
[217] "T" "L" "V" "S" "T" "N" "K" "R" "S" "A" "Q" "R" "S" "S" "L" "T" "P" "A"
[235] "S" "S" "S" "A" "G" "V" "K" "R" "S" "K" "T" "S" "T" "S" "A" "N" "V" "M"
[253] "D" "P" "L" "D" "S" "T" "T" "E" "T" "G" "A" "T" "T" "T" "A" "Q" "L" "V"
[271] "P" "Q" "Q" "I" "T" "V" "Q" "T" "S" "V" "V" "S" "A" "A" "E" "A" "K" "L"
[289] "H" "Q" "Q" "S" "P" "Q" "Q" "V" "R" "Q" "E" "E" "A" "E" "Y" "I" "D" "L"
[307] "P" "M" "E" "L" "P" "T" "K" "S" "E" "P" "D" "Y" "S" "E" "D" "H" "G" "D"
[325] "A" "A" "G" "D" "A" "E" "G" "T" "Y" "V" "E" "D" "D" "T" "Y" "G" "D" "M"
[343] "R" "Y" "D" "D" "S" "Y" "F" "T" "E" "N" "E" "D" "A" "G" "N" "Q" "T" "A"
[361] "A" "N" "T" "S" "G" "G" "G" "V" "T" "A" "T" "T" "S" "K" "A" "V" "V" "K"
[379] "Q" "Q" "S" "Q" "N" "Y" "S" "E" "S" "S" "F" "V" "D" "T" "S" "G" "D" "Q"
[397] "G" "N" "T" "E" "A" "Q" "V" "T" "Q" "H" "V" "R" "N" "C" "G" "P" "Q" "M"
[415] "F" "L" "I" "S" "R" "K" "G" "G" "T" "L" "L" "T" "I" "N" "N" "F" "V" "Y"
[433] "R" "S" "N" "L" "K" "F" "F" "G" "K" "S" "N" "N" "I" "L" "Y" "W" "E" "C"
[451] "V" "Q" "N" "R" "S" "V" "K" "C" "R" "S" "R" "L" "K" "T" "I" "G" "D" "D"
[469] "L" "Y" "V" "T" "N" "D" "V" "H" "N" "H" "M" "G" "D" "N" "K" "R" "I" "E"
[487] "A" "A" "K" "A" "A" "G" "M" "L" "I" "H" "K" "K" "L" "S" "S" "L" "T" "A"
[505] "A" "D" "K" "I" "Q" "G" "S" "W" "K" "M" "D" "T" "E" "G" "N" "P" "D" "H"
[523] "L" "P" "K" "M" "*"
\end{Soutput}
\end{Schunk}
\subsection{\texttt{getType()}}
\begin{Schunk}
\begin{Sinput}
 # Need internet connection
   choosebank("emblTP")
   getType()
\end{Sinput}
\begin{Soutput}
        sname                                  libel
2661      CDS              .PE protein coding region
2662       ID                            Locus entry
2663 MISC_RNA .RN other structural RNA coding region
2664     RRNA          .RR Ribosomal RNA coding gene
2665    SCRNA              .SC small cytoplasmic RNA
2666    SNRNA                  .SN small nuclear RNA
2667     TRNA           .TR Transfer RNA coding gene
\end{Soutput}
\end{Schunk}
\subsection{\texttt{getlistrank()}}
\begin{Schunk}
\begin{Sinput}
  # Need internet connection
  choosebank("emblTP")
  query("MyListName", "sp=Borrelia burgdorferi", virtual = TRUE)
  (result <- getlistrank("MyListName"))
\end{Sinput}
\begin{Soutput}
[1] 2
\end{Soutput}
\begin{Sinput}
  stopifnot(result == 2)
  closebank()
\end{Sinput}
\end{Schunk}
\subsection{\texttt{getliststate()}}
\begin{Schunk}
\begin{Sinput}
   ### Need internet connection
   choosebank("emblTP")
   query("mylist", "sp=felis catus et t=cds", virtual=TRUE)
   getliststate(glr("mylist")) # SQ, MYLIST, 603, FALSE
\end{Sinput}
\begin{Soutput}
$type
[1] "SQ"

$name
[1] "MYLIST"

$count
[1] 603

$locus
[1] FALSE
\end{Soutput}
\begin{Sinput}
   gln(glr("mylist")) # MYLIST (upper case letters on server)
\end{Sinput}
\begin{Soutput}
[1] "MYLIST"
\end{Soutput}
\begin{Sinput}
   closebank()
\end{Sinput}
\end{Schunk}
\subsection{\texttt{gfrag()}}
\begin{Schunk}
\begin{Sinput}
 # Need internet connection
   choosebank("emblTP")
   gfrag("LMFLCHR36", start = 1, length = 3529852) -> myseq
   stopifnot(nchar(myseq) == 3529852)
   closebank()   
\end{Sinput}
\end{Schunk}
\subsection{\texttt{ghelp()}}
\begin{Schunk}
\begin{Sinput}
   ### Need internet connection
   choosebank("emblTP")
   ghelp()
\end{Sinput}
\begin{Soutput}
      ---- General Information on ACNUC nucleic acid data base ----             
HELP:                                                                           
   A detailed explanation of purpose and usage of each command is obtained      
by typing the command name and requesting help when the dialog suggests it.     
SEQUENCES AND SUBSEQUENCES:                                                     
   In addition to sequences as published in research articles, ACNUC contains   
subsequences which are sequence segments with specific coding function (e.g.    
protein, tRNA, rRNA genes...). Sequence type distinguishes parent from sub-     
sequences: parent sequences have ID type, subsequences have a type that      
indicates their function (CDS, TRNA, RRNA,...). Most subsequence names derive   
from the parent sequence's name by addition of suffixes .PEn, .TRn, .RRn, .SNn  
.RNn for CDS, TRNA, RRNA, snRNA or misc_RNA-typed subsequences, respectively.   
When the gene name is known, it is used as a suffix in the corresponding        
subsequence name.
SEQUENCE LISTS:                                                                 
   This program deals with sequence lists which group sequences selected from   
the data base using one or more selection criteria (see SELECT help). Many      
sequence lists can be handled simultaneously by the program and previous lists  
can be used to define new ones.                                                 
   Typical use of program is:                                                   
 - SPECIES command to know which species names are to be used in selection.     
 - KEYWORDS command to know which keywords are to be used in selection.         
 - SELECT command to select sequences from data base combining various          
criteria. This command produces the list of sequences that fit the criteria.    
 - SHORT command to obtain a brief description of selected sequences            
or - INFO command to get more detailed information.                             
 - EXTRACT command to copy selected sequences to a user file.                   
LIST NAMES:                                                                     
   Lists are created by commands SELECT or FIND. They are given automatically a 
name (LIST1, LIST2,...) by the program, unless the user enters his own list     
name by appending /l=my_list_name to the command name at the "Command?" prompt. 
Most commands operate either on a sequence list or on an individual sequence.   
Reply to question "List, sequence, or accession #? [default=...]" with          
<RETURN> to access the default (list of) sequence(s) or with any list name,     
sequence name, or accession number.                                             
FILE OUTPUT:                                                                    
If /lpt is appended to command name at the "Command?" prompt, the output of     
commands SPECIES, KEYWORDS, INFO, SHORT, NAMES, CODES, BASES goes to a file     
named `query.out'.                                                              
CODED NAMES:                                                                    
   Coded names are to be used when specifying species, keywords, journals,      
sequence types, organelles, molecules. Specific commands    
(SPECIES, KEYWORDS, CODES) allow you to find these names easily.                
REFERENCES:                                                                     
   To find a sequence from a bibliographical reference use the selection        
criterion "R=reference-code" of SELECT command. Build the reference code as     
follows (journal names are given by CODES command):                             
  journal_name/volume/first_page         for journal articles                   
  book/year/name_of_1st_author           for books                              
  thesis/year/name_of_1st_author         for thesis                             
  patent/patent_number                   for patented sequences                 
  unpubl/year/name_of_1st_author         for unpublished sequences              
Example: nar/8/2173 stands for Nucleic Acids Research 8:2173-2192 (1980).       
\end{Soutput}
\begin{Sinput}
   ghelp("SELECT")
\end{Sinput}
\begin{Soutput}
     In addition to functions described in the help for the simple usage of     
command SELECT, other selection criteria and operations between lists exist.    
Specifically, it is also possible to build lists of species and lists of        
keywords for further retrieval capabilities.                                    
                                                                                
Criteria      Resulting selection                                               
FK=file name  List of keywords taken from a file (which may have been created   
              by a SAVE command).                                               
FS=file name  List of species taken from a file (which may have been created    
              by a SAVE command).                                               
                                                                                
Operation     Result                                                            
ME list       Replaces subsequences in list by sequences from which they        
              are extracted (equivalent to option 4 of command MODIFY).         
FI list       Sequences in list plus all of their subsequences (equivalent to   
              option 5 of command MODIFY).                                      
PS list       Produces the list of species names attached to sequences in list. 
PK list       Produces the list of keyword names attached to sequences in list. 
UN list       If applied to a species list, produces the list of sequences from 
              species in the list; if applied to a keyword list, produces the   
              list of sequences attached to keywords in list.                   
SD spec-list  Applied to a list of species, produces the list of all descendants
              from them in the species tree. The list itself can easily be      
              created by command FIND.                                          
KD keyw-list  Applied to a list of keywords, produces the list of all           
              descendants from them in the keywords tree. The list itself can   
              easily be created by command FIND.                                
                                                                                
Operators PS, PK, and UN allow to solve the problem "find all genes             
simultaneously sequenced in a given series of species".                         
First, build the lists of sequences from each of these species. Next project    
each of these lists to attached keywords by applying operator PK. Then compute  
the list of keywords in common by combining the keyword lists with operator ET. 
Then, remove from this list of common keywords, those which are                 
uncharacteristic (e.g. partial) by employing command MODIFY. Finally, produce   
the lists of sequences attached to common keywords from each species by         
applying operator UN combined with initial species-based sequence lists.        
Species and keyword lists can be listed with command NAMES and saved with SAVE. 
\end{Soutput}
\begin{Sinput}
   # To get info about current database:
   ghelp("CONT")
\end{Sinput}
\begin{Soutput}
             ****     ACNUC Data Base Content      ****                         
              EMBL Library Release 78 WITHOUT ESTs  (March 2004)
27,571,397,913 bases; 12,533,594 sequences; 1,604,500 subseqs; 339,186 refers.
Software by M. Gouy & M. Jacobzone, Laboratoire de biometrie, Universite Lyon I 
\end{Soutput}
\end{Schunk}
\subsection{\texttt{isenum()}}
\begin{Schunk}
\begin{Sinput}
   ### Need internet connection
   choosebank("emblTP")
   isenum("LMFLCHR36")
\end{Sinput}
\begin{Soutput}
$number
[1] 13682678

$length
[1] 3529852

$frame
[1] 0

$gencode
[1] 0

$ncbigc
[1] 1

$otheraccessmatches
[1] FALSE
\end{Soutput}
\begin{Sinput}
   isn("LMFLCHR36")
\end{Sinput}
\begin{Soutput}
[1] 13682678
\end{Soutput}
\begin{Sinput}
   stopifnot(isn("LMFLCHR36") == 13682678)
   # Example with CDS:
   isenum("AB004237")
\end{Sinput}
\begin{Soutput}
$number
[1] 66351

$length
[1] 1140

$frame
[1] 0

$gencode
[1] 2

$ncbigc
[1] 2

$otheraccessmatches
[1] FALSE
\end{Soutput}
\end{Schunk}
\subsection{\texttt{knowndbs()}}
\begin{Schunk}
\begin{Sinput}
   ### Need internet connection
   choosebank("emblTP")
   kdb()
\end{Sinput}
\begin{Soutput}
          bank status
1      genbank     on
2         embl     on
3      emblwgs     on
4    swissprot     on
5      ensembl     on
6       refseq     on
7        nrsub     on
8    hobacnucl     on
9    hobacprot     on
10 hovergendna     on
11    hovergen     on
12     hogenom     on
13  hogenomdna     on
14   hogennucl     on
15   hogenprot     on
16   hoverclnu     on
17   hoverclpr     on
18    homolens     on
19 homolensdna     on
20     greview     on
21 polymorphix     on
22      emglib     on
23   HAMAPnucl     on
24   HAMAPprot     on
25   hoppsigen     on
26   nurebnucl     on
27   nurebprot     on
28  taxobacgen     on
                                                                            info
1                   GenBank Rel. 167 (15 August 2008) Last Updated: Oct 26, 2008
2            EMBL Library Release 96 (September 2008) Last Updated: Oct 25, 2008
3               EMBL Whole Genome Shotgun sequences Release 96  (September 2008)
4        UniProt Rel. 14 (SWISS-PROT 56 + TrEMBL 39): Last Updated: Aug 28, 2008
5                       Ensembl Release 49\t     \t\t Last Updated: Apr 23, 2008
6                        RefSeq 15.0 (1 January 2006) Last Updated: Jan 23, 2006
7                                    NRSub database release 10.1 (December 1997)
8                        HOBACGEN - genomic data - Release 10 (February 12 2002)
9                        HOBACGEN - protein data - Release 10 (February 12 2002)
10 HOVERGEN - genomic data - Release 48 (May 24 2007) Last Updated: May 24, 2007
11 HOVERGEN - protein data - Release 48 (May 24 2007) Last Updated: May 24, 2007
12 HOGENOM - protein data - Release 04 (Sept 18,2007) Last Updated: Feb 27, 2008
13 HOGENOM - genomic data - Release 04 (Sept 18,2007) Last Updated: Feb 21, 2008
14  HOGENOM - genomic data - Release 03 (Oct 14 2005) Last Updated: Nov  7, 2005
15  HOGENOM - protein data - Release 03 (Oct 14 2005) Last Updated: Mar 10, 2006
16                      HOVERGEN CLEAN - genomic data - Release 46 (Jun 10 2004)
17                      HOVERGEN CLEAN - protein data - Release 46 (Jun 10 2004)
18   HOMOLENS 4 - Homologous genes from Ensembl(49)\t Last Updated: Jul  4, 2008
19    HOMOLENS 4 - Homologous genes from Ensembl(49)\tLast Updated: Jul  4, 2008
20              EBI Genome Reviews. Acnuc Release 2. Last Updated: June 19, 2005
21                                         POLYBASE - Release 1  (June 20, 2003)
22                                           EMGLib Release 5 (December 9, 2003)
23                                                                   HAMAP nucl.
24                                                                   HAMAP prot.
25                                                                     Hoppsigen
26                   Nurebase 4.0 (26 September 2003) Last Updated: NOV 27, 2003
27                   Nurebase 4.0 (26 September 2003) Last Updated: NOV 27, 2003
28                                            TaxoBacGen Rel. 7 (September 2005)
\end{Soutput}
\begin{Sinput}
   closebank()
\end{Sinput}
\end{Schunk}
\subsection{\texttt{oriloc()}}
\begin{Schunk}
\begin{Sinput}
 #
 # A little bit too long for routine checks because oriloc() is already
 # called in draw.oriloc.Rd documentation file. Try example(draw.oriloc)
 # instead, or copy/paste the following code:
 #
 out <- oriloc()
 plot(out$st, out$sk, type = "l", xlab = "Map position in Kb",
     ylab = "Cumulated composite skew", 
     main = expression(italic(Chlamydia~~trachomatis)~~complete~~genome))
 #
 # Example with a single GenBank file:
 #
 out2 <- oriloc(gbk=system.file("sequences/ct.gbk", package = "seqinr"))
 draw.oriloc(out2)
\end{Sinput}
\end{Schunk}
\includegraphics{../figs/rundontrun-oriloc}
\subsection{\texttt{prepgatannots()}}
\begin{Schunk}
\begin{Sinput}
 # Need internet connection
   choosebank("genbank")
   query("mylist","n=AQF16SRRN")
   pga() # We want to scan all annotations, including FEATURES
   modifylist("mylist", operation = "strain", type = "scan")
   mylist$nelem # should be 1
\end{Sinput}
\begin{Soutput}
[1] 1
\end{Soutput}
\end{Schunk}
\subsection{\texttt{prettyseq()}}
\begin{Schunk}
\begin{Sinput}
   ### Need internet connection
   choosebank("emblTP")
   prettyseq(111)
\end{Sinput}
\begin{Soutput}
Name: A00165           Length:108
Genetic code used: NUG=AUN=M when initiation codon

                  10         20         30         40         50         60 
           Q  Y  C   G  N  L  S   T  C  M   L  G  T   Y  T  Q  D   F  N  K
          cagtactgcg gtaatctgag tacttgcatg ctgggcacat acacgcagga cttcaacaag 
          >A00165

                  70         80         90        100        110 
           F  H  T   F  P  Q  T   A  I  G   V  G  A   P  G  *
          tttcacacgt tcccccaaac tgcaattggg gttggagcac ctggttga   
                                                       A00165<
\end{Soutput}
\end{Schunk}
\subsection{\texttt{print.SeqAcnucWeb()}}
\begin{Schunk}
\begin{Sinput}
   ### Need internet connection
   choosebank("emblTP")
   query("mylist", "sp=felis catus")     
   mylist$req[[1]]
\end{Sinput}
\begin{Soutput}
    name   length    frame   ncbicg 
"A06937"     "34"      "0"      "1" 
\end{Soutput}
\end{Schunk}
\subsection{\texttt{print.qaw()}}
\begin{Schunk}
\begin{Sinput}
   ### Need internet connection
   choosebank("emblTP")
   query("sp=felis catus")       
   list1
\end{Sinput}
\begin{Soutput}
4732 SQ for sp=felis catus
\end{Soutput}
\end{Schunk}
\subsection{\texttt{query()}}
\begin{Schunk}
\begin{Sinput}
  # Need internet connection
  choosebank("genbank")
  query("bb", "sp=Borrelia burgdorferi")
  # To get the names of the 4 first sequences:
  sapply(bb$req[1:4], getName)
\end{Sinput}
\begin{Soutput}
[1] "A04009" "A22442" "A24006" "A24008"
\end{Soutput}
\begin{Sinput}
  # To get the 4 first sequences:
  sapply(bb$req[1:4], getSequence, as.string = TRUE) 
\end{Sinput}
\begin{Soutput}
[1] "aagcttaattagaaccaaacttaattaaaaccaaacttaattgaagttattatcattttattttttttcaattttctatttgttatttgttaatcttataatataattatacttgtattaagttatattaatataaaaggagaatatattatgaaaaaatatttattgggaataggtctaatattagccttaatagcatgtaagcaaaatgttagcagccttgacgagaaaaacagcgtttcagtagatttgcctggtgaaatgaaagttcttgtaagcaaagaaaaaaacaaagacggcaagtacgatctaattgcaacagtagacaagcttgagcttaaaggaacttctgataaaaacaatggatctggagtacttgaaggcgtaaaagctgacaaaagtaaagtaaaattaacaatttctgacgatctaggtcaaaccacacttgaagttttcaaagaagatggcaaaacactagtatcaaaaaaagtaacttccaaagacaagtcatcaacagaagaaaaattcaatgaaaaaggtgaagtatctgaaaaaataataacaagagcagacggaaccagacttgaatacacaggaattaaaagcgatggatctggaaaagctaaagaggttttaaaaggctatgttcttgaaggaactctaactgctgaaaaaacaacattggtggttaaagaaggaactgttactttaagcaaaaatatttcaaaatctggggaagtttcagttgaacttaatgacactgacagtagtgctgctactaaaaaaactgcagcttggaattcaggcacttcaactttaacaattactgtaaacagtaaaaaaactaaagaccttgtgtttacaaaagaaaacacaattacagtacaacaatacgactcaaatggcaccaaattagaggggtcagcagttgaaattacaaaacttgatgaaattaaaaacgctttaaaataaggagaatttatgagattattaataggatttgctttagcgttagctttaataggatgtgcacaaaaaggtgctgagtcaattggttctcaaaaagaaaatgatctaaaccttgaagactctagtaaaaaatcacatcaaaacgctaaacaagaccttcctgcggtgacagaagactcagtgtctttgtttaatggtaataaaatttttgtaagcaaagaaaaaaatagctccggcaaatatgatttaagagcaacaattgatcaggttgaacttaaaggaacttccgataaaaacaatggttctggaacccttgaaggttcaaagcctgacaagagtaaagtaaaattaacagtttctgctgatttaaacacagtaaccttagaagcatttgatgccagcaaccaaaaaatttcaagtaaagttactaaaaaacaggggtcaataacagaggaaactctcaaagctaataaattagactcaaagaaattaacaagatcaaacggaactacacttgaatactcacaaataacagatgctgacaatgctacaaaagcagtagaaactctaaaaaatagcattaagcttgaaggaagtcttgtagtcggaaaaacaacagtggaaattaaagaaggtactgttactctaaaaagagaaattgaaaaagatggaaaagtaaaagtctttttgaatgacactgcaggttctaacaaaaaaacaggtaaatgggaagacagtactagcactttaacaattagtgctgacagcaaaaaaactaaagatttggtgttcttaacagatggtacaattacagtacaacaatacaacacagctggaaccagcctagaaggatcagcaagtgaaattaaaaatctttcagagcttaaaaacgctttaaaataatatataagtaaaccccctacaaggcatcagctagtgtaggaag"
[2] "atgaaaaaatatttattgggaataggtctaatattagccttaatagcatgtaagcaaaatgttagcagccttgacgagaaaaacagcgtttcagtagatttgcctggtgaaatgaacgttcttgtaagcaaagaaaaaaacaaagacggcaagtacgatctaattgcaacagtagacaagcttgagcttaaaggaacttctgataaaaacaatggatctggagtacttgaaggcgtaaaagctgacaaaagtaaagtaaaattaacaatttctgacgatctaggtcaaaccacacttgaagttttcaaagaagatggcaaaacactagtatcaaaaaaagtaacttccaaagacaagtcatcaacagaagaaaaattcaatgaaaaaggtgaagtatctgaaaaaataataacaagagcagacggaaccagacttgaatacacagaaattaaaagcgatggatctggaaaagctaaagaggttttaaaaagctatgttcttgaaggaactttaactgctgaaaaaacaacattggtggttaaagaaggaactgttactttaagcaaaaatatttcaaaatctggggaagtttcagttgaacttaatgacactgacagtagtgctgctactaaaaaaactgcagcttggaattcaggcacttcaactttaacaattactgtaaacagtaaaaaaactaaagaccttgtgtttacaaaagaaaacacaattacagtacaacaatacgactcaaatggcaccaaattagaggggtcagcagttgaaattacaaaacttgatgaaattaaaaacgctttaaaataa"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
[3] "atgaaaaaatatttattgggaataggtctaatattagccttaatagcatgtaagcaaaatgttagcagccttgatgaaaaaaatagcgtttcagtagatttacctggtggaatgaaagttcttgtaagtaaagaaaaagacaaagatggtaaatacagtctagaggcaacagtagacaagcttgagcttaaaggaacttctgataaaaacaacggttctggaacacttgaaggtgaaaaaactgacaaaagtaaagtaaaattaacaattgctgaggatctaagtaaaaccacatttgaaattttcaaagaagatggcaaaacattagtatcaaaaaaagtaacccttaaagacaagtcatcaacagaagaaaaattcaacgaaaagggtgaaatatctgaaaaaacaatagtaagagcaaatggaaccagacttgaatacacagacataaaaagcgatggatccggaaaagctaaagaagttttaaaagactttactcttgaaggaactctagctgctgacggcaaaacaacattgaaagttacagaaggcactgttgttttaagcaagaacattttaaaatccggagaaataacagttgcacttgatgactctgacactactcaggctactaaaaaaactggaaaatgggattcaaagacttccactttaacaattagtgtgaatagccaaaaaaccaaaaaccttgtattcacaaaagaagacacaataacagtacaaaaatacgactcagcaggcaccaatctagaaggcaaagcagtcgaaattacaacacttaaagaacttaaagacgctttaaaataa"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
[4] "atgaaaaaatatttattgggaataggtctaatattagccttaatagcatgtaagcaaaatgttagcagccttgacgagaaaaacagcgtttcagtagatgtacctggtggaatgaaagttcttgtaagcaaagaaaaaaacaaagacggcaagtacgatctaatggcaacagtggacaacgttgatcttaaaggaacttctgacaaaaacaatggatctggaatacttgaaggcgtaaaagctgataaaagtaaagtaaaattaacagttgctgacgatctaagcaaaaccacacttgaagttttaaaagaagatggtacagtagtgtcaagaaaagtaacttccaaagacaagtcaacaacagaagcaaaattcaacgaaaaaggtgaattgtctgaaaaaacaatgacaagagcaaacggaactactcttgaatactcacaaatgacaaatgaagacaatgctgcaaaagcagtagaaactcttaaaaacggcattaagtttgaaggaaatctcgcaagtggaaaaacagcagtggaaattaaagaaggcactgttactctaaaaagagaaattgataaaaatggaaaagtaaccgtctctttaaatgacactgcatctggttctaaaaaaacagcttcctggcaagaaagtactagcaccttaacaattagtgcaaacagcaaaaaaactaaagatctagtgttcctaacaaacggtacaattacagtacaaaattatgactcagctggcactaaacttgaaggatcagcagctgaaattaaaaaactcgatgaacttaaaaacgctttaagataa"                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        
\end{Soutput}
\end{Schunk}
\subsection{\texttt{readfirstrec()}}
\begin{Schunk}
\begin{Sinput}
 # Need internet connection
   choosebank("genbank")
   allowedtype <- readfirstrec()
   sapply(allowedtype, function(x) readfirstrec(type = x))
\end{Sinput}
\begin{Soutput}
      AUT       BIB       ACC       SMJ       SUB       LOC       KEY 
   174467    525954  96474557      5351 102053262  99964643   9391212 
     SPEC      SHRT       LNG       EXT       TXT 
   598454 894561842  28901625   8301690    477859 
\end{Soutput}
\end{Schunk}
\subsection{\texttt{rearranged.oriloc()}}
\begin{Schunk}
\begin{Sinput}
 r.ori <- rearranged.oriloc(seq.fasta = system.file("sequences/ct.fasta",package = "seqinr"),
     g2.coord = system.file("sequences/ct.coord",package = "seqinr"))
\end{Sinput}
\end{Schunk}
\subsection{\texttt{residuecount()}}
\begin{Schunk}
\begin{Sinput}
   ### Need internet connection
   choosebank("emblTP")
   query("mylist", "t=CDS", virtual = TRUE)
   stopifnot(residuecount(glr("mylist")) == 1611439240)
   stopifnot(is.na(residuecount(glr("unknowlist")))) # A warning is issued
\end{Sinput}
\end{Schunk}
\subsection{\texttt{savelist()}}
\begin{Schunk}
\begin{Sinput}
   ### Need internet connection
   choosebank("emblTP")
   query("mylist", "sp=felis catus et t=cds", virtual=TRUE)
   savelist(glr("mylist"))
\end{Sinput}
\begin{Soutput}
603 sequence mnemonics written into file: MYLIST.mne
\end{Soutput}
\begin{Sinput}
   # 603 sequence mnemonics written into file: MYLIST.mne 
   savelist(glr("mylist"), type = "A")
\end{Sinput}
\begin{Soutput}
603 sequence accession numbers written into file: MYLIST.acc
\end{Soutput}
\end{Schunk}
\subsection{\texttt{setlistname()}}
\begin{Schunk}
\begin{Sinput}
   ### Need internet connection
   choosebank("emblTP")
   query("mylist", "sp=felis catus et t=CDS", virtual = TRUE)
   # Change list name on server:
   setlistname(lrank = glr("mylist"), name = "feliscatus") # 0, OK.
\end{Sinput}
\begin{Soutput}
[1] 0
\end{Soutput}
\begin{Sinput}
   glr("mylist") # 0, list doesn't exist no more.
\end{Sinput}
\begin{Soutput}
[1] 0
\end{Soutput}
\begin{Sinput}
   glr("feliscatus") # 2, this list exists.
\end{Sinput}
\begin{Soutput}
[1] 2
\end{Soutput}
\end{Schunk}
\subsection{\texttt{translate()}}
\begin{Schunk}
\begin{Sinput}
 ## Need internet connection.
 ## Translation of the following EMBL entry:
 ##
 ## FT   CDS             join(complement(153944..154157),complement(153727..153866),
 ## FT                   complement(152185..153037),138523..138735,138795..138955)
 ## FT                   /codon_start=1
 ## FT                   /db_xref="FLYBASE:FBgn0002781"
 ## FT                   /db_xref="GOA:Q86B86"
 ## FT                   /db_xref="TrEMBL:Q86B86"
 ## FT                   /note="mod(mdg4) gene product from transcript CG32491-RZ;
 ## FT                   trans splicing"
 ## FT                   /gene="mod(mdg4)"
 ## FT                   /product="CG32491-PZ"
 ## FT                   /locus_tag="CG32491"
 ## FT                   /protein_id="AAO41581.1"
 ## FT                   /translation="MADDEQFSLCWNNFNTNLSAGFHESLCRGDLVDVSLAAEGQIVKA
 ## FT                   HRLVLSVCSPFFRKMFTQMPSNTHAIVFLNNVSHSALKDLIQFMYCGEVNVKQDALPAF
 ## FT                   ISTAESLQIKGLTDNDPAPQPPQESSPPPAAPHVQQQQIPAQRVQRQQPRASARYKIET
 ## FT                   VDDGLGDEKQSTTQIVIQTTAAPQATIVQQQQPQQAAQQIQSQQLQTGTTTTATLVSTN
 ## FT                   KRSAQRSSLTPASSSAGVKRSKTSTSANVMDPLDSTTETGATTTAQLVPQQITVQTSVV
 ## FT                   SAAEAKLHQQSPQQVRQEEAEYIDLPMELPTKSEPDYSEDHGDAAGDAEGTYVEDDTYG
 ## FT                   DMRYDDSYFTENEDAGNQTAANTSGGGVTATTSKAVVKQQSQNYSESSFVDTSGDQGNT
 ## FT                   EAQVTQHVRNCGPQMFLISRKGGTLLTINNFVYRSNLKFFGKSNNILYWECVQNRSVKC
 ## FT                   RSRLKTIGDDLYVTNDVHNHMGDNKRIEAAKAAGMLIHKKLSSLTAADKIQGSWKMDTE
 ## FT                   GNPDHLPKM"
 choosebank("emblTP")
 query("trans", "N=AE003734.PE35")
 getTrans(trans$req[[1]])
\end{Sinput}
\begin{Soutput}
  [1] "M" "A" "D" "D" "E" "Q" "F" "S" "L" "C" "W" "N" "N" "F" "N" "T" "N" "L"
 [19] "S" "A" "G" "F" "H" "E" "S" "L" "C" "R" "G" "D" "L" "V" "D" "V" "S" "L"
 [37] "A" "A" "E" "G" "Q" "I" "V" "K" "A" "H" "R" "L" "V" "L" "S" "V" "C" "S"
 [55] "P" "F" "F" "R" "K" "M" "F" "T" "Q" "M" "P" "S" "N" "T" "H" "A" "I" "V"
 [73] "F" "L" "N" "N" "V" "S" "H" "S" "A" "L" "K" "D" "L" "I" "Q" "F" "M" "Y"
 [91] "C" "G" "E" "V" "N" "V" "K" "Q" "D" "A" "L" "P" "A" "F" "I" "S" "T" "A"
[109] "E" "S" "L" "Q" "I" "K" "G" "L" "T" "D" "N" "D" "P" "A" "P" "Q" "P" "P"
[127] "Q" "E" "S" "S" "P" "P" "P" "A" "A" "P" "H" "V" "Q" "Q" "Q" "Q" "I" "P"
[145] "A" "Q" "R" "V" "Q" "R" "Q" "Q" "P" "R" "A" "S" "A" "R" "Y" "K" "I" "E"
[163] "T" "V" "D" "D" "G" "L" "G" "D" "E" "K" "Q" "S" "T" "T" "Q" "I" "V" "I"
[181] "Q" "T" "T" "A" "A" "P" "Q" "A" "T" "I" "V" "Q" "Q" "Q" "Q" "P" "Q" "Q"
[199] "A" "A" "Q" "Q" "I" "Q" "S" "Q" "Q" "L" "Q" "T" "G" "T" "T" "T" "T" "A"
[217] "T" "L" "V" "S" "T" "N" "K" "R" "S" "A" "Q" "R" "S" "S" "L" "T" "P" "A"
[235] "S" "S" "S" "A" "G" "V" "K" "R" "S" "K" "T" "S" "T" "S" "A" "N" "V" "M"
[253] "D" "P" "L" "D" "S" "T" "T" "E" "T" "G" "A" "T" "T" "T" "A" "Q" "L" "V"
[271] "P" "Q" "Q" "I" "T" "V" "Q" "T" "S" "V" "V" "S" "A" "A" "E" "A" "K" "L"
[289] "H" "Q" "Q" "S" "P" "Q" "Q" "V" "R" "Q" "E" "E" "A" "E" "Y" "I" "D" "L"
[307] "P" "M" "E" "L" "P" "T" "K" "S" "E" "P" "D" "Y" "S" "E" "D" "H" "G" "D"
[325] "A" "A" "G" "D" "A" "E" "G" "T" "Y" "V" "E" "D" "D" "T" "Y" "G" "D" "M"
[343] "R" "Y" "D" "D" "S" "Y" "F" "T" "E" "N" "E" "D" "A" "G" "N" "Q" "T" "A"
[361] "A" "N" "T" "S" "G" "G" "G" "V" "T" "A" "T" "T" "S" "K" "A" "V" "V" "K"
[379] "Q" "Q" "S" "Q" "N" "Y" "S" "E" "S" "S" "F" "V" "D" "T" "S" "G" "D" "Q"
[397] "G" "N" "T" "E" "A" "Q" "V" "T" "Q" "H" "V" "R" "N" "C" "G" "P" "Q" "M"
[415] "F" "L" "I" "S" "R" "K" "G" "G" "T" "L" "L" "T" "I" "N" "N" "F" "V" "Y"
[433] "R" "S" "N" "L" "K" "F" "F" "G" "K" "S" "N" "N" "I" "L" "Y" "W" "E" "C"
[451] "V" "Q" "N" "R" "S" "V" "K" "C" "R" "S" "R" "L" "K" "T" "I" "G" "D" "D"
[469] "L" "Y" "V" "T" "N" "D" "V" "H" "N" "H" "M" "G" "D" "N" "K" "R" "I" "E"
[487] "A" "A" "K" "A" "A" "G" "M" "L" "I" "H" "K" "K" "L" "S" "S" "L" "T" "A"
[505] "A" "D" "K" "I" "Q" "G" "S" "W" "K" "M" "D" "T" "E" "G" "N" "P" "D" "H"
[523] "L" "P" "K" "M" "*"
\end{Soutput}
\end{Schunk}



\section*{Session Informations}

This part was compiled under the following \Rlogo{}~environment:

\begin{itemize}
  \item R version 2.8.0 (2008-10-20), \verb|i386-apple-darwin8.8.2|
  \item Locale: \verb|fr_FR.UTF-8/fr_FR.UTF-8/fr_FR.UTF-8/C/C/C|
  \item Base packages: base, datasets, grDevices, graphics, methods,
    stats, utils
  \item Other packages: MASS~7.2-44, ade4~1.4-9, ape~2.2-2,
    nlme~3.1-89, quadprog~1.4-11, seqinr~2.0-0, tseries~0.10-16,
    xtable~1.5-4, zoo~1.5-4
  \item Loaded via a namespace (and not attached): grid~2.8.0,
    lattice~0.17-15
\end{itemize}
There were two compilation steps:

\begin{itemize}
  \item \Rlogo{} compilation time was: Sun Oct 26 18:44:17 2008
  \item \LaTeX{} compilation time was: \today
\end{itemize}

% END - DO NOT REMOVE THIS LINE

%%%%%%%%%%%%  BIBLIOGRAPHY %%%%%%%%%%%%%%%%%
\clearpage
\addcontentsline{toc}{section}{References}
\bibliographystyle{plain}
\bibliography{../config/book}
\end{document}
